---
title: "scATACseq_R-SKMK_ReportCreateQuality"
author: "Laëtitia Racine"
date: "2022-09-22"
subtitle: "Last modification : `r format(Sys.time(), '%d %B, %Y')`"
output:
  html_document:
    code_folding: "hide"
    toc: true
    toc_float: false
    theme: journal
---

<style>
body {text-align: justify}
</style>

```{r, Setup, include=F}

knitr::opts_chunk$set(echo = TRUE)
knitr::opts_chunk$set(warning = FALSE)
knitr::opts_chunk$set(message = FALSE)

```

```{r, Dependencies}

library(knitr)
library(Seurat)
library(Signac)
library(stringr)
library(gridExtra)
library(kableExtra)

```

```{r, Working directory }

start_time = Sys.time()
directory = str_extract(string = getwd(), pattern = "[:graph:]+(?=bin)")
directory_skmk = paste0(directory, "exp/", "scATACseq_SKMK_preprocessing/")

```


# scATACseq_SKMK_CreateSeuratObject - Report

## Script command lines to create Seurat Object

```{r, eval = FALSE}

#**************
# Input loading
#**************

matrixh5 = arguments$matrixh5_file
fragment = arguments$fragment_file
singlecell = arguments$metadata_file

name_cond = str_extract(string = singlecell, pattern = "[:alnum:]+(?=_singlecell.csv)")


#*********************
# Create Seurat Object
#*********************

counts <- Seurat::Read10X_h5(filename = matrixh5)

metadata <- read.csv(
  file = singlecell,
  header = TRUE,
  row.names = 1
)

chrom_assay <- CreateChromatinAssay(
  counts = counts,
  sep = c(":", "-"),
  genome = 'hg38',
  fragments = fragment,
  min.cells = 100,
  min.features = 200
)

seurat_obj <- CreateSeuratObject(
  counts = chrom_assay,
  assay = "peaks",
  meta.data = metadata)


#*****************
# Add annotations
#*****************

# Extract annotations from EnsDb database
annotations <- GetGRangesFromEnsDb(ensdb = EnsDb.Hsapiens.v86, standard.chromosomes = TRUE)
seqlevelsStyle(annotations) = "UCSC"
print(table(annotations$type))

# Add annotation database into seurat object (needed for downstream analysis (QC functions))
Annotation(seurat_obj) = annotations

# Visualize first object of the list to see the annotations
seurat_obj@assays$peaks@annotation


#*****************
# Update orig.ident
#*****************

Idents(object = seurat_obj) = name_cond
seurat_obj$orig.ident = name_cond


#************
# Save output
#************

saveRDS(object = seurat_obj, file = arguments$output)

```

## CTRL object visualisation

```{r, eval = TRUE, results = 'hold'}

# Load file 
seurat_obj = readRDS(paste0(directory_skmk, "CTRL_seurat.rds"))

# Global structure
seurat_obj
## chromatin assay with the peaks
seurat_obj@assays$peaks
## list of peaks from chromatin assay
seurat_obj@assays$peaks@ranges
## raw count matrix store in counts.slot (first 10 rows and column)
as.data.frame(seurat_obj@assays$peaks@counts)[1:10, 1:10]
## normalized count matrix store in data.slot (first 10 rows and column)
as.data.frame(seurat_obj@assays$peaks@data)[1:10, 1:10]
# => same as the raw matrix because the normalization is not done yet

# database annotations
seurat_obj@assays$peaks@annotation

```


# scATACseq_SKMK_QualityControl - Report

https://caleblareau.com/cshl-ast2019-scATAC/code/scATAC-tutorial.html  
https://satijalab.org/signac/articles/pbmc_vignette.html  
https://satijalab.org/signac/articles/mouse_brain_vignette.html  
  
Metrics to check to remove outliers cell if needed :     
- Nucleosome banding pattern : fragment length periodicity, the histogram of DNA fragment sizes should exhibit a strong nucleosome banding pattern corresponding to the length of DNA wrapped around a single nucleosome.    
- Ratio reads in genomic blacklist regions : Cells with a high proportion of reads mapping to these areas often represent technical artifacts and should be removed.       
- Fraction of fragments in peaks   
- Total number of fragments : the total number of high-quality fragments captured for the cell, Cells with very few reads may need to be excluded due to low sequencing depth. Cells with extremely high levels may represent doublets, nuclei clumps, or other artefacts.   
- Transcriptional start site enrichment score : Poor ATAC-seq experiments typically will have a low TSS enrichment score.     

First, we visualize the initial number of  cells. 
Then, we compute QC metrics and filter the data.

NB : nucleosome signal limit at 2, possible to adapt depending the data  
We try to find the most fitted limits for all conditions = to apply same filtering on all data  

Grange = liste des peaks avec leur location génomique (chr, start, end)
Count = matrice de count avec la liste des peaks et le nombre de fragments qu'ils contiennent par cellule
Attention lorsqu'on change le grange, ça ne change pas automatiquement la matrice de count. 
Il faut ensuite faire correspondre les listes de peak pour ne garder que ceux qui sont dans le grange filtré. 

## Script command lines

```{r, eval = FALSE}

#**************
# Input loading
#**************

seurat_obj = readRDS(file = arguments$seurat_obj)
name_cond = arguments$cond

  
#************************
# Initial number of cells
#************************

# Extract number of cells and peaks 
nbcells = ncol(seurat_obj)
nbpeaks = nrow(seurat_obj)

# Store data in a tab
df_summary = data.frame("Conditions" = name_cond,
                        "Initial_Cell_Number" = nbcells,
                        "Initial_Peak_Number" = nbpeaks)


#***************************
# Nucleosome banding pattern
#***************************

seurat_obj = NucleosomeSignal(object = seurat_obj)
seurat_obj$nucleosome_group = ifelse(seurat_obj$nucleosome_signal > "2", 'NS > 2', 'NS < 2')
  
qc_plot_nucl_fragments = FragmentHistogram(object = seurat_obj) +
  geom_vline(xintercept = 147, color = "red") +  # no nucleosome limit
  geom_vline(xintercept = 294, color = "red") +  # one nucleosome limit
  geom_vline(xintercept = 441, color = "red") +  # two nucleosomes limit
  geom_vline(xintercept = 588, color = "red") +  # three nucleosomes limit
  labs(title = "Nucleosome banding pattern") +
  theme(title = element_text(face = "bold"))
  
qc_plot_nucl_signal = VlnPlot(
  object = seurat_obj,
  features = 'nucleosome_signal',
  pt.size = 0.1) + 
  theme(legend.position = 'none',
        axis.title.x = element_blank(),
        axis.text.x = element_blank()) +  
  geom_hline(yintercept = 2, color = "red") 
  
qc_plot_nucl_quality = FragmentHistogram(
    object = seurat_obj, 
    group.by = 'nucleosome_group') +
    labs(title = "Nucleosome signal") +
    theme(title = element_text(face = "bold"))


#**********************************
# Number of fragments and blacklist
#**********************************

seurat_obj$pct_reads_in_peaks = seurat_obj$peak_region_fragments / 
    seurat_obj$passed_filters * 100
seurat_obj$blacklist_ratio = seurat_obj$blacklist_region_fragments /
    seurat_obj$peak_region_fragments
  
qc_plot_pct_reads_vln = VlnPlot(object = seurat_obj,
                                features = 'pct_reads_in_peaks',
                                pt.size = 0.1) + 
    theme(legend.position = 'none',
          axis.title.x = element_blank(),
          axis.text.x = element_blank())
  
qc_plot_pct_reads_point = qplot(log10(seurat_obj$passed_filters),
                                  seurat_obj$pct_reads_in_peaks) +
    labs(x = "log10 # of fragments", y = "% reads in peaks") +
    geom_hline(yintercept = 60, color = "firebrick") +
    geom_vline(xintercept = log10(5000), color = "firebrick")
  
qc_plot_blacklist = VlnPlot(object = seurat_obj,
                            features = 'blacklist_ratio',
                            pt.size = 0.1) + 
    theme(legend.position = 'none',
          axis.title.x = element_blank(),
          axis.text.x = element_blank())
  
qc_plot_peak_region_fragments = VlnPlot(object = seurat_obj,
                                        features = 'peak_region_fragments',
                                        pt.size = 0.1) + 
    theme(legend.position = 'none',
          axis.title.x = element_blank(),
          axis.text.x = element_blank())


#*********************
# TSS enrichment score 
#*********************

seurat_obj = TSSEnrichment(object = seurat_obj, fast = FALSE)
seurat_obj$high.tss = ifelse(seurat_obj$TSS.enrichment > 2, 'High', 'Low')
seurat_obj$pct_reads_in_TSS = seurat_obj$TSS_fragments / seurat_obj$passed_filters * 100
  
qc_plot_tss = TSSPlot(seurat_obj, group.by = 'high.tss') + 
    NoLegend() +
    labs(title = "TSS enrichment score") +
    theme(title = element_text(face = "bold"))
  
qc_plot_tss_reads =  qplot(seurat_obj$pct_reads_in_TSS, seurat_obj$pct_reads_in_peaks) +
    labs(x = "% reads in TSS", y = "% reads in peaks") +
    geom_abline(intercept = 0, slope = 1, linetype = 2, color = "firebrick")
  
 
#*************************************
# Visualize QC metrics and save plots
#*************************************

list_graphs = list(qc_plot_nucl_fragments, qc_plot_nucl_signal, qc_plot_nucl_quality,
              qc_plot_pct_reads_vln, qc_plot_pct_reads_point, qc_plot_blacklist,
              qc_plot_peak_region_fragments, qc_plot_tss, qc_plot_tss_reads)
ggsave(plot = arrangeGrob(grobs = list_graphs, 
                          ncol = 3, 
                          nrow = 3,
                          top = textGrob(paste("QC metrics for", name_cond))),
         filename = arguments$output_plots_panel,
         width = 40, height = 40)

saveRDS(object = list_graphs, file = arguments$output_plots_list)
saveRDS(object = seurat_obj, file = arguments$output_seurat)


#**************************************************************
# Number of cells after QC filtering  and remove outlier cells
#**************************************************************

# Initialize summary df with initial numbers of cells and peaks
vec_cell = c(paste0(name_cond, "_NbCells"))
vec_peak = c(paste0(name_cond, "_NbPeaks"))
df_filter = data.frame(
  "Condition" = c(vec_cell, vec_peak),
  "Initial" = c(df_summary$Initial_Cell_Number, df_summary$Initial_Peak_Number))

# Apply first filter on nucleosome signal 
seurat_obj = subset(x = seurat_obj, subset = nucleosome_signal < 2)
df_filter$Nucleosome_filter = c(ncol(seurat_obj), nrow(seurat_obj))

# Apply filter on number of fragments in cell
seurat_obj =  subset(x = seurat_obj, subset = peak_region_fragments > 3000 & 
                       peak_region_fragments < 50000)
df_filter$Fragments_filter =  c(ncol(seurat_obj), nrow(seurat_obj))

# Apply filter on percentage of fragments in peaks
seurat_obj =  subset(x = seurat_obj, subset = pct_reads_in_peaks > 50)
df_filter$PctReads_filter =  c(ncol(seurat_obj), nrow(seurat_obj))

# Apply filter on blacklist fragments
seurat_obj =  subset(x = seurat_obj, subset = blacklist_ratio < 0.05)
df_filter$Blacklist_filter =  c(ncol(seurat_obj), nrow(seurat_obj))

# Apply filter on TSS enrichment score
seurat_obj =  subset(x = seurat_obj, subset = TSS.enrichment > 2)
df_filter$TSS_filter =  c(ncol(seurat_obj), nrow(seurat_obj))

## Clean chromosome list - Keep only standard chromosomes
seurat_obj@assays$peaks@ranges = tidyChromosomes(gr = seurat_obj@assays$peaks@ranges,
                                                 keep.X = TRUE,
                                                 keep.Y = TRUE,
                                                 keep.M = FALSE,
                                                 keep.nonstandard = FALSE,
                                                 genome = "hg38")
list_peak_clean = seurat_obj@assays$peaks@ranges$peak_name
seurat_obj = subset(x = seurat_obj, features = list_peak_clean) 
df_filter$Chromosome_filter = c(ncol(seurat_obj), nrow(seurat_obj))

#**********************
# Save filtered output
#**********************

saveRDS(object = seurat_obj, file = arguments$output_seurat_filtered)
write.table(x = df_filter, file = arguments$output_tab)




```

## Filtering summary for all conditions

```{r}

# Load summary tab for each condition
CTRL_df_filter = read.csv2(file = paste0(directory_skmk, "CTRL_df_filter.csv"), sep = "")
DDG_df_filter = read.csv2(file = paste0(directory_skmk, "2DG_df_filter.csv"), sep = "")
DON_df_filter = read.csv2(file = paste0(directory_skmk, "DON_df_filter.csv"), sep = "")
AOA_df_filter = read.csv2(file = paste0(directory_skmk, "AOA_df_filter.csv"), sep = "")

# Merge tab into one global tab
df_filter_glob = rbind(CTRL_df_filter, DDG_df_filter, DON_df_filter, AOA_df_filter)

df_filter_glob %>%
  kable() %>%
  kable_styling()

```


```{r, Rsession}

end_time = Sys.time()
cat("Total execution time : ", as.numeric (end_time - start_time, units = "mins"), "minutes")

# Clean working space and memory 
rm(list = ls())
gc()

# Show package version
sessionInfo()
```
