---
title: "scATACseq_R_DWNS_CompareConditions_DA"
author: "LaÃ«titia Racine"
date: "2023-04-13"
subtitle: "Last modification : `r format(Sys.time(), '%d %B, %Y')`"
output:
  html_document:
    code_folding: "hide"
    toc: true
    toc_float: false
    theme: journal
---

<style>
body {text-align: justify}
</style>

```{r, Setup, include=FALSE}

knitr::opts_chunk$set(echo = TRUE)
knitr::opts_chunk$set(warning = FALSE)
knitr::opts_chunk$set(message = FALSE)

```

```{r, Dependencies}

library(Seurat)
library(Signac)
library(knitr)
library(stringr)
library(ggplot2)
library(dplyr)
library(kableExtra)
library(GenomicRanges)
library(tibble)
library(pheatmap)
# motif analysis
library(JASPAR2020)
library(TFBSTools)
library(BSgenome.Hsapiens.UCSC.hg38)
library(gridExtra)
# GO analysis
library(org.Hs.eg.db)
library(clusterProfiler)

```

```{r, Working directories and external script}

directory = str_extract(string = getwd(), pattern = "[:graph:]+(?=bin)")
start_time = Sys.time()

# Create a unique folder for output corresponding to the date of the day
current_date = format(Sys.time(), "%Y%m%d")
dir.create(path = paste0(directory,"exp/","scATACseq_R_DWNS_CompareConditions_DA/"))
dir.create(path = paste0(directory,"exp/","scATACseq_R_DWNS_CompareConditions_DA/", 
                         current_date))
directory_output = paste0(directory,"exp/","scATACseq_R_DWNS_CompareConditions_DA/", 
                          current_date, "/")

# Load external script with functions and constants
source(file = paste0(directory, "bin/", "functions_constants.R"))

```

```{r, Input loading}

dir = pic_last_dir(paste0(directory, "exp/", "scATACseq_R_NormalizationReduction/"))
seurat_obj = readRDS(paste0(dir, "/", "merged_seurat_qc_filter_annot_normreduc.rds"))

```



<br><br><br>



# Overview 

See scATACseq_R_DWNS_ClustersInvestigation.Rmd for more information. Same code here but with conditions as groups instead of clusters.

<br>

As a reminder : 

```{r, fig.width=16, fig.height=4}

seurat_obj = SetIdent(object = seurat_obj, value = "orig.ident")
ind_umap = list() 

for (i in 1:length(unique(seurat_obj$orig.ident))) {
  
  # Select condition to highlight
  highlight_cond = unique(seurat_obj$orig.ident)[i]
  color = unname(noquote(color_code[highlight_cond]))
  others_cond = setdiff(unique(seurat_obj$orig.ident), highlight_cond)
  others_cond_group = paste(others_cond, collapse=", ")
  
  # Sort cell according to their condition
  cell_highlight = unlist(CellsByIdentities(object = seurat_obj, idents = highlight_cond))
  cell_nohighlight = unlist(CellsByIdentities(object = seurat_obj, idents = others_cond))
  
  # Plot 
  plot_umap = DimPlot(object = seurat_obj,
                      group.by = "orig.ident",
                      reduction ="umap",
                      pt.size = 0.1,
                      cells.highlight = list(cell_highlight, cell_nohighlight),
                      cols.highlight = c(color,"grey")) +
    ggtitle(highlight_cond) +
    theme(legend.position = "none")
  
  ind_umap[[i]] = plot_umap
  
}

clust_umap = DimPlot(seurat_obj, 
                     reduction = "umap", 
                     group.by = "peaks_snn_res.0.25") + 
  ggtitle("Resolution_0.25") 

clust_umap | ind_umap[[1]] | ind_umap[[2]] | ind_umap[[3]] | ind_umap[[4]]

```



<br><br><br>


# Extract list of DA peaks for each condition compare to all others

We used the FindAllMarkers function to extract differentially accessible peaks for each condition. This function is really slow so we run it on IFB external cluster to have more performance. We used the slurm job on long partition with memory 100GB accessible. https://satijalab.org/seurat/reference/findallmarkers    
```{r, Markers FAM on IFB, eval=FALSE}

DefaultAssay(seurat_obj) <- 'peaks'
Idents(seurat_obj) = "orig.ident"

cond_markers = FindAllMarkers(object = seurat_obj,
                              assay = "peaks",
                              features = NULL, # use all genes
                              logfcthreshold = 0.25,
                              test.use = "LR", # not default
                              slot = "data",
                              min.pct = 0.05, # not default
                              min.diff.pct = -Inf,
                              node = NULL,
                              verbose = TRUE,
                              only.pos = TRUE, # to keep only positives markers
                              max.cells.per.ident = Inf,
                              random.seed = 42, # not default
                              latent.vars = NULL,
                              min.cells.feature = 3,
                              min.cells.group = 3,
                              mean.fxn = NULL,
                              fc.name = NULL,
                              base = 2,
                              return.thresh = 2, # to remove threshold on pvalue
                              densify = FALSE)

saveRDS(object = cond_markers, paste0(directory_output, "cond_markers_FAM_",
                              current_date, ".rds"))

```

```{r}

# Load the cond_markers table from IFBcluster
cond_markers = readRDS(paste0(directory, "data/scATACseq/cond_markers_FAM_20230416.rds"))
cat("Number of DA peaks for each condition : \n")
table(cond_markers$cluster)

cond_markers = cond_markers %>%
  dplyr::rename("peak_name" = "gene") %>%
  dplyr::rename("condition" = "cluster") %>%
  tidyr::separate(col = "peak_name", into = c("seqnames", "start", "end"), remove=FALSE)

saveRDS(cond_markers, paste0(directory_output, "cond_markers.rds"))

```

<br>

We tried in parallel to use FindMarkers manually and we compared the results. We obtained the same results (it all depends of the value of return.thresh parameter in FindAllMarkers). The advantage of this method is that there are intermediate savings. If something happens between each conditions calculation, we still have the previous results saved whereas with FindAllMarkers, it's more automatic but we loose everything.   https://satijalab.org/seurat/reference/findmarkers      
```{r, Markers FM on IFB, eval=FALSE}

DefaultAssay(seurat_obj) <- 'peaks'
Idents(seurat_obj) = "orig.ident"

cond_markers_CTRL = FindMarkers(object = seurat_obj,
                                ident.1 = "CTRL",
                                ident.2 = c("DON", "2DG", "AOA"),
                                group.by = NULL,
                                subset.ident = NULL,
                                assay = "peaks",
                                slot = "data",
                                reduction = NULL,
                                features = NULL, #use all genes
                                logfc.threshold = 0.25,
                                test.use = "LR",
                                min.pct = 0.05,
                                min.diff.pct = -Inf,
                                verbose = TRUE,
                                only.pos = TRUE,
                                max.cells.per.ident = Inf,
                                random.seed = 42,
                                latent.vars = NULL,
                                min.cells.feature = 3,
                                min.cells.group = 3,
                                mean.fxn = NULL,
                                fc.name = NULL,
                                base = 2,
                                densify = FALSE)

cond_markers_CTRL$condition = "CTRL"
saveRDS(cond_markers_CTRL, paste0(directory_output,
                                  "cond_markers_CTRL_",
                                  current_date,
                                  ".rds"))

cond_markers_DON = FindMarkers(object = seurat_obj,
                               ident.1 = "DON",
                               ident.2 = c("CTRL", "2DG", "AOA"),
                               group.by = NULL,
                               subset.ident = NULL,
                               assay = "peaks",
                               slot = "data",
                               reduction = NULL,
                               features = NULL, #use all genes
                               logfc.threshold = 0.25,
                               test.use = "LR",
                               min.pct = 0.05,
                               min.diff.pct = -Inf,
                               verbose = TRUE,
                               only.pos = TRUE,
                               max.cells.per.ident = Inf,
                               random.seed = 42,
                               latent.vars = NULL,
                               min.cells.feature = 3,
                               min.cells.group = 3,
                               mean.fxn = NULL,
                               fc.name = NULL,
                               base = 2,
                               densify = FALSE)
cond_markers_DON$condition = "DON"
saveRDS(cond_markers_DON, paste0(directory_output,
                                 "cond_markers_DON_",
                                 current_date, ".rds"))

cond_markers_2DG = FindMarkers(object = seurat_obj,
                               ident.1 = "2DG",
                               ident.2 = c("CTRL", "DON", "AOA"),
                               group.by = NULL,
                               subset.ident = NULL,
                               assay = "peaks",
                               slot = "data",
                               reduction = NULL,
                               features = NULL, #use all genes
                               logfc.threshold = 0.25,
                               test.use = "LR",
                               min.pct = 0.05,
                               min.diff.pct = -Inf,
                               verbose = TRUE,
                               only.pos = TRUE,
                               max.cells.per.ident = Inf,
                               random.seed = 42,
                               latent.vars = NULL,
                               min.cells.feature = 3,
                               min.cells.group = 3,
                               mean.fxn = NULL,
                               fc.name = NULL,
                               base = 2,
                               densify = FALSE)

cond_markers_2DG$condition = "2DG"
saveRDS(cond_markers_2DG, paste0(directory_output,
                                 "cond_markers_2DG_",
                                 current_date, ".rds"))

cond_markers_AOA = FindMarkers(object = seurat_obj,
                               ident.1 = "AOA",
                               ident.2 = c("CTRL", "2DG", "DON"),
                               group.by = NULL,
                               subset.ident = NULL,
                               assay = "peaks",
                               slot = "data",
                               reduction = NULL,
                               features = NULL, #use all genes
                               logfc.threshold = 0.25,
                               test.use = "LR",
                               min.pct = 0.05,
                               min.diff.pct = -Inf,
                               verbose = TRUE,
                               only.pos = TRUE,
                               max.cells.per.ident = Inf,
                               random.seed = 42,
                               latent.vars = NULL,
                               min.cells.feature = 3,
                               min.cells.group = 3,
                               mean.fxn = NULL,
                               fc.name = NULL,
                               base = 2,
                               densify = FALSE)

cond_markers_AOA$condition = "AOA"
saveRDS(cond_markers_AOA, paste0(directory_output,
                                 "cond_markers_AOA_",
                                 current_date, ".rds"))

cond_markers = rbind(cond_markers_CTRL, cond_markers_DON,
                     cond_markers_AOA, cond_markers_2DG)

cond_markers = cond_markers %>%
  tibble::rownames_to_column("peak_name") %>%
  tidyr::separate(col = "peak_name", into = c("seqnames", "start", "end"), remove=FALSE)

saveRDS(cond_markers, paste0(directory_output, "cond_markers.rds"))

```

<br>

See branch FAM_FM_IFBcluster on git project to access .R and .sh scripts used for this part.    

```{r, results="hide"}

list_to_keep = c("directory_output", "start_time", "seurat_obj", 
                 "cond_markers", "pic_last_dir", "directory")
rm(list = setdiff(ls(), list_to_keep))
gc()

```



<br><br><br>



# Associated gene for each peak

```{r, results="hide"}

# Extract localisation of hg38 annotation (see annotationsFiles_createGrange.Rmd)
dir = pic_last_dir(paste0(directory, "exp/AnnotationsFiles_createGrange/"))
annot_hg38 = readRDS(paste0(dir, "/", "hg38_annotations_gr.rds"))

annot_df = data.frame(annot_hg38) %>%
  dplyr::select(seqnames, start, end, gene_id, symbol, type) 
annot_df = annot_df %>%
  dplyr::mutate(gene_id = ifelse(is.na(gene_id), type, gene_id)) %>%
  dplyr::mutate(symbol = ifelse(is.na(symbol), type, symbol))

rm(annot_hg38)
gc()

```

```{r}

# transform tables into granges
annot_gr = makeGRangesFromDataFrame(annot_df, keep.extra.columns = TRUE)
cond_gr = makeGRangesFromDataFrame(cond_markers, keep.extra.columns = TRUE)

# find overlaps between granges 
overlaps = findOverlaps(query = annot_gr, subject = cond_gr, ignore.strand = TRUE)

# add corresponding columns from query granges into subject granges
mcols(cond_gr)$gene_id = NA
mcols(cond_gr)$symbol = NA
mcols(cond_gr)$type = NA
mcols(cond_gr)$gene_id[subjectHits(overlaps)] <- mcols(annot_gr)$gene_id[queryHits(overlaps)]
mcols(cond_gr)$symbol[subjectHits(overlaps)] <- mcols(annot_gr)$symbol[queryHits(overlaps)]
mcols(cond_gr)$type[subjectHits(overlaps)] <- mcols(annot_gr)$type[queryHits(overlaps)]

# final table
tab_cond = data.frame(cond_gr)
table(tab_cond$type, tab_cond$condition)

# save 
write.csv2(tab_cond, paste0(directory_output, "tab_cond_DApeaks_annotation.csv"), row.names = FALSE)

```

```{r, results="hide"}

list_to_keep = c("directory_output", "start_time", "seurat_obj", 
                 "tab_cond", "directory")
rm(list = setdiff(ls(), list_to_keep))
gc()

```



<br><br><br>



# Select relevant peaks in the list (top 20) to plot on heatmap

```{r, Select relevant markers}

# filter on p value adjusted
cat("We remove markers with p_val_adj superior to 0.05.")
tab_cond_relevant = tab_cond %>% dplyr::filter(p_val_adj < 0.05)

# Extract top 20 of positive markers based on avg_log2FC only 
cat("We keep the positive markers with the biggest average_log2FC for each condition.")
top20_pos = tab_cond_relevant %>%
  dplyr::filter(avg_log2FC >= 0) %>%
  dplyr::group_by(condition) %>%
  top_n(n=20, wt = avg_log2FC) %>%
  dplyr::select(condition,symbol, p_val_adj, everything()) %>%
  ungroup()

# positive markers : calcul the difference between pct1 and pct2
cat("We filter the positive markers to keep those with a difference > to 0.1 between pct.1 and pct.2.", "\n",
    "Note : we don't use the same threshold on diff as in cluster investigation because the difference between conditions are more less important.")
top20_pos_filtered = tab_cond_relevant %>% 
  dplyr::filter(avg_log2FC >= 0) %>%
  dplyr::mutate(diff = pct.1 - pct.2) %>%
  dplyr::filter(diff > 0.1) %>%
  dplyr::group_by(condition) %>%
  top_n(n=20, wt = avg_log2FC) %>%
  dplyr::select(condition, symbol, p_val_adj, everything()) %>%
  dplyr::select(-diff) %>%
  ungroup()

# Summary markers
summary_cond = data.frame(condition = c("2DG", "AOA", "CTRL", "DON"), 
                           nb_pos_markers = as.vector(table(top20_pos$condition)),
                           nb_pos_markers_filtered = as.vector(table(top20_pos_filtered$condition)))

summary_cond %>%
  kable() %>%
  kable_styling()

# save 
write.csv2(top20_pos, paste0(directory_output, "top20_pos.csv"), row.names = FALSE)
write.csv2(top20_pos_filtered, paste0(directory_output, "top20_pos_filtered.csv"), row.names = FALSE)

```

<br>

https://davetang.org/muse/2018/05/15/making-a-heatmap-in-r-with-the-pheatmap-package/

```{r, Heatmap,  fig.height = 26, fig.width = 14}

# Randomly downsample 200 cells of each condition
cells_df = as.data.frame(seurat_obj@meta.data)  %>%
  tibble::rownames_to_column("cell") %>%
  dplyr::select(cell, orig.ident) %>%
  dplyr::rename("condition" = "orig.ident") 
# table(cells_df$condition)

cells_df = cells_df %>%
  dplyr::group_by(condition) %>%
  dplyr::sample_n(200) %>%
  tibble::column_to_rownames("cell")

cells_to_keep = rownames(cells_df)

# Extract names of peaks to keep
peaks_df = top20_pos_filtered %>% 
  dplyr::select(symbol, peak_name) %>%
  dplyr::distinct()
peaks_to_keep = peaks_df$peak_name

# Filter matrix to keep only wanted cells and peaks
mat = as.data.frame(seurat_obj@assays$peaks@data[peaks_to_keep, cells_to_keep])
mat = mat %>% tibble::rownames_to_column("peak_name")
mat = left_join(mat, peaks_df, by = "peak_name") 

mat_ht = mat %>% 
  dplyr::mutate(peak_name = paste0(symbol, " - ", peak_name)) %>%
  dplyr::select(-symbol) %>% 
  tibble:::column_to_rownames("peak_name") 

# scale the rows
cal_z_score <- function(x){(x - mean(x)) / sd(x)}
mat_ht <- t(apply(mat_ht, 1, cal_z_score))

# Draw plot
ht_all = pheatmap(mat = mat_ht, 
                  annotation_col = cells_df,
                  show_colnames = FALSE,
                  cluster_cols = FALSE,
                  cluster_rows = FALSE,
                  fontsize_rows = 6)

mat_filt = mat %>%
  dplyr::filter(!symbol %in% str_subset(symbol, pattern = "hg38")) %>% 
  dplyr::mutate(peak_name = paste0(symbol, " - ", peak_name)) %>%
  dplyr::select(-symbol) %>% 
  tibble:::column_to_rownames("peak_name") 
mat_filt <- t(apply(mat_filt, 1, cal_z_score))

ht_filt = pheatmap(mat = mat_filt, 
                   annotation_col = cells_df,
                   show_colnames = FALSE,
                   cluster_cols = FALSE,
                   cluster_rows = FALSE,
                   fontsize_rows = 6)

```

```{r, results="hide"}

save_pheatmap_png <- function(x, filename, width=1200, height=1200, res = 150) {
  png(filename, width = width, height = height, res = res)
  grid::grid.newpage()
  grid::grid.draw(x$gtable)
  dev.off()
}
 
save_pheatmap_png(ht_all, paste0(directory_output, "heatmap_DApeaks_cond_all.png"))
save_pheatmap_png(ht_filt, paste0(directory_output, "heatmap_DApeaks_cond_genes.png"))

```

```{r, results="hide"}

list_to_keep = c("directory_output", "start_time", "seurat_obj", 
                 "tab_cond_relevant", "directory")
rm(list = setdiff(ls(), list_to_keep))
gc()

```



<br><br><br>



# GO on list of peaks differentially accessible between conditions

```{r, fig.show = "hide"}

markers = tab_cond_relevant %>% dplyr::select(condition, symbol, gene_id)
markers = markers %>% dplyr::filter(!symbol %in% str_subset(symbol, "hg38"))

list_GO_analysis = list()

for (i in 1:length(unique(tab_cond_relevant$condition))) {
  
  cond = as.character(unique(tab_cond_relevant$condition)[i])
  markers_cond = (markers %>% dplyr::filter(condition == cond))$gene_id

  ego = enrichGO(gene = markers_cond,
                 OrgDb = "org.Hs.eg.db",
                 ont = "ALL",
                 pAdjustMethod = "BH",
                 pvalueCutoff = 0.05,
                 qvalueCutoff = 0.2,
                 readable = TRUE,
                 keyType = "ENTREZID")

  ego_df = ego@result %>% dplyr::mutate(condition = cond)
  
  if (nrow(ego_df) !=0) { 
    
    plot = dotplot(ego, showCategory = 25) +
      facet_grid(ONTOLOGY ~ ., scales="free") +
      ggtitle(paste0("DApeaks - Closest genes enrichment - ", cond)) +
      theme(axis.text.y = element_text(size = 8))
    
    list_GO_analysis[[cond]]= list(dotplot = plot,
                                   ego = ego_df)
  
  } else {
      
    list_GO_analysis[[cond]] = list(ego = ego_df)
    
    }

}

saveRDS(list_GO_analysis, paste0(directory_output, "list_GO_cond.rds"))

```

```{r, fig.width=14, fig.height=20}

list = unlist(list_GO_analysis, recursive = FALSE)
list_dotplot = list[str_subset(names(list), "dotplot")]
do.call("grid.arrange", c(list_dotplot, ncol = 2))

list_tab = list[str_subset(names(list), "ego")]
tab_ego = do.call("rbind", list_tab)
tab_ego %>%
  kable() %>%
  kable_styling() %>%
  scroll_box(width = "100%", height="250px")

```

```{r, fig.show="hide"}

ggsave(plot = do.call("grid.arrange", c(list_dotplot, ncol=2)), width=22,
       height=30, filename=paste0(directory_output, "GO_DApeaks_cond.svg"))

write.csv2(tab_ego, paste0(directory_output, "tab_GO_DApeaks_cond.csv"), row.names=FALSE)

```

```{r, results="hide"}

list_to_keep = c("directory_output", "start_time", "seurat_obj", 
                 "tab_cond_relevant", "directory")
rm(list = setdiff(ls(), list_to_keep))
gc()

```



<br><br><br>



# Motif analysis on list of peaks differentially accessible between conditions

Documentation :     
https://stuartlab.org/signac/articles/motif_vignette.html    
https://www.rdocumentation.org/packages/TFBSTools/versions/1.10.3/topics/getMatrixSet    
https://stuartlab.org/signac/reference/addmotifs    

```{r}

# Get a list of motif position frequency matrices from the JASPAR database
pfm <- getMatrixSet(
  x = JASPAR2020,
  opts = list(collection = "CORE", species = "Homo sapiens", 
              tax_group = 'vertebrates', all_versions = FALSE)
)

# add motif information
seurat_obj <- AddMotifs(
  object = seurat_obj,
  genome = BSgenome.Hsapiens.UCSC.hg38,
  pfm = pfm
)

```

<br>

This part of the code with ChromVar is quite long. We can use the same object as in scATACseq_R_DWNS_ClustersInvestigation_DA.Rmd to go faster.  
```{r}

# Compute a per-cell motif activity score !!extremely slow!!
seurat_obj <- RunChromVAR(
  object = seurat_obj,
  genome = BSgenome.Hsapiens.UCSC.hg38
)

saveRDS(seurat_obj, paste0(directory_output, "seurat_motif.rds"))

```

```{r, eval=FALSE}

# To gain time
seurat_obj = readRDS(paste0(directory, "exp/scATACseq_R_DWNS_ClustersInvestigation_DA/seurat_motif.rds"))

```

<br>

Search for DNA motifs that are overrepresented in a set of peaks that are differentially accessible between conditions.  
```{r}

list_enrich_cond = list()

for (i in 1:length(unique(tab_cond_relevant$condition))) {
  
  cond = as.character(unique(tab_cond_relevant$condition)[i])
  
  # extract list of peaks
  peaks_cond = tab_cond_relevant %>% dplyr::filter(condition == cond)
  peaks_cond = peaks_cond$peak_name

  # test enrichment
  enriched.motifs <- FindMotifs(object = seurat_obj,
                                features = peaks_cond)
  enriched.motifs$condition = cond
    
  # plot more enriched motifs
  plot = MotifPlot(object = seurat_obj,
                   motifs = head(rownames(enriched.motifs))) +
    ggtitle(cond)
  
  # save in a list 
  list_enrich_cond[[cond]] = list(tab_motif = enriched.motifs,
                                  plot_motif = plot)
  
}

```

```{r, fig.width=18, fig.height=10}

list = unlist(list_enrich_cond, recursive = FALSE)
list_tab = list[str_detect(names(list), "tab")]
list_plot = list[str_detect(names(list), "plot")]

glob_enrich_tab = do.call("rbind", list_tab)
glob_enrich_tab[1:10,] %>%
  kable(caption = "First 10 rows of the table") %>%
  kable_styling() %>%
  scroll_box(width = "100%", height = "250px")

do.call("grid.arrange", c(list_plot, ncol = 2))

```

```{r, results="hide", fig.show="hide"}

ggsave(plot = do.call("grid.arrange", c(list_plot, ncol = 2)),
       filename = paste0(directory_output, "enriched_motifs_cond_plot.png"),
       width = 20, height = 12)

write.csv2(glob_enrich_tab, paste0(directory_output, "enriched_motifs_cond_tab.csv"))

```

<br>

Visualize motif activities per cell and provide an alternative method of identifying differentially-active motifs between conditions. 
```{r}

# identify interesting and/or common motifs

motif_df = glob_enrich_tab %>% dplyr::filter(p.adjust < 0.01)
common_motif_df = motif_df[motif_df$motif %in% motif_df$motif[duplicated(motif_df$motif)],]
common_motif_list = unique(common_motif_df$motif)
unique_motif_df = motif_df[!motif_df$motif %in% motif_df$motif[duplicated(motif_df$motif)],]
unique_motif_list = unique(unique_motif_df$motif)
  
# Display table 

common_motif_df %>%
  kable(caption = "Enriched motif in conditions with padj < 0.01 - shared motifs") %>%
  kable_styling() %>%
  scroll_box(width = "100%", height = "250px")
  
unique_motif_df %>%
  kable(caption = "Enriched motif in conditions with padj < 0.01 - unique motifs") %>%
  kable_styling() %>%
  scroll_box(width = "100%", height = "250px")

# Plot the results

DefaultAssay(seurat_obj) <- 'chromvar'

unique_motif_plot = list()
common_motif_plot = list()

for(i in 1:length(unique_motif_list)){
  unique_motif_plot[[i]] = FeaturePlot(
    object = seurat_obj,
    features = unique_motif_list[i],
    min.cutoff = 'q10',    
    max.cutoff = 'q90',
    pt.size = 0.1)
}

for(j in 1:length(common_motif_list)){
  common_motif_plot[[j]] = FeaturePlot(
    object = seurat_obj,
    features = common_motif_list[j],
    min.cutoff = 'q10',    
    max.cutoff = 'q90',
    pt.size = 0.1)
}

```

```{r, fig.width=40, fig.height=45, fig.show="hold"}

cat("Motifs enriched in only one condition.")
do.call("grid.arrange", c(unique_motif_plot[1:32], ncol = 6))
do.call("grid.arrange", c(unique_motif_plot[33:65], ncol = 6))
do.call("grid.arrange", c(unique_motif_plot[66:98], ncol = 6))

```

```{r, fig.width=40, fig.height=45, fig.show="hold"}

cat("Sample of motifs enriched in multiple conditions.")
do.call("grid.arrange", c(common_motif_plot[1:36], ncol = 6))
do.call("grid.arrange", c(common_motif_plot[37:74], ncol = 6))

```

```{r, fig.show="hide"}

write.csv2(unique_motif_df, paste0(directory_output, "unique_enriched_motif_cond.csv"))
write.csv2(common_motif_df, paste0(directory_output, "common_enriched_motif_cond.csv"))

ggsave(plot= do.call("grid.arrange", c(unique_motif_plot[1:32], ncol=6)), width=40,
       height=40, filename=paste0(directory_output, "unique_enriched_motif_cond_plot1.png"))
ggsave(plot= do.call("grid.arrange", c(unique_motif_plot[33:65], ncol=6)), width=40,
       height=40, filename=paste0(directory_output, "unique_enriched_motif_cond_plot2.png"))
ggsave(plot= do.call("grid.arrange", c(unique_motif_plot[66:98], ncol=6)), width=40,
       height=40, filename=paste0(directory_output, "unique_enriched_motif_cond_plot3.png"))

ggsave(plot= do.call("grid.arrange", c(common_motif_plot[1:32], ncol=6)), width=40,
       height=40, filename=paste0(directory_output, "common_enriched_motif_cond_plot1.png"))
ggsave(plot= do.call("grid.arrange", c(common_motif_plot[33:65], ncol=6)), width=40,
       height=40, filename=paste0(directory_output, "common_enriched_motif_cond_plot2.png"))
ggsave(plot= do.call("grid.arrange", c(common_motif_plot[66:98], ncol=6)), width=40,
       height=40, filename=paste0(directory_output, "common_enriched_motif_cond_plot3.png"))
ggsave(plot= do.call("grid.arrange", c(common_motif_plot[99:131], ncol=6)), width=40,
       height=40, filename=paste0(directory_output, "common_enriched_motif_cond_plot4.png"))
ggsave(plot= do.call("grid.arrange", c(common_motif_plot[132:160], ncol=6)), width=40,
       height=40, filename=paste0(directory_output, "common_enriched_motif_cond_plot5.png"))

```



<br><br><br>



```{r, Rsession}

end_time = Sys.time()
cat("Total execution time : ", as.numeric (end_time - start_time, units = "mins"), "minutes")

# Show package version
sessionInfo()

```

```{r, results = "hide"}

# Clean working space and memory 
rm(list = ls())
gc()

```
