---
title: "scATACseq_R_MergeObjectsCommonPeaks"
author: "Laëtitia Racine"
date: "2022-09-02"
subtitle: "Last modification : `r format(Sys.time(), '%d %B, %Y')`"
output:
  html_document:
    code_folding: "hide"
    toc: true
    toc_float: false
    theme: journal
---

<style>
body {text-align: justify}
</style>

```{r, Setup, include=FALSE}

knitr::opts_chunk$set(echo = TRUE)
knitr::opts_chunk$set(warning = FALSE)
knitr::opts_chunk$set(message = FALSE)

```

```{r, Dependencies}

library(Seurat)
library(Signac)
library(GenomicRanges) # function width()
library(stringr)

```

```{r, Working directories and external script}

directory = str_extract(string = getwd(), pattern = "[:graph:]+(?=bin)")
directory_data = paste0(directory, "data/scATACseq/")
start_time = Sys.time()

# Create a unique folder for output corresponding to the date of the day
current_date = format(Sys.time(), "%Y%m%d")
dir.create(path = paste0(directory,"exp/","scATACseq_R_MergeObjectsCommonPeaks/"))
dir.create(path = paste0(directory,"exp/","scATACseq_R_MergeObjectsCommonPeaks/", current_date))
directory_output = paste0(directory,"exp/","scATACseq_R_MergeObjectsCommonPeaks/", current_date, "/")

# Load external script with functions and constants
source(file = paste0(directory, "bin/", "functions_constants.R"))

```

```{r, Input loading}

dir = paste0(directory, "exp/", "scATACseq_SKMK_preprocessing/")
list_seurat_obj = list("CTRL" = readRDS(paste0(dir, "/", "CTRL_seurat_qc_filter.rds")),
                       "2DG" = readRDS(paste0(dir, "/", "2DG_seurat_qc_filter.rds")),
                       "DON" = readRDS(paste0(dir, "/", "DON_seurat_qc_filter.rds")),
                       "AOA" = readRDS(paste0(dir, "/", "AOA_seurat_qc_filter.rds")))

# List of files available in data folder
fragment_list = list.files(path = directory_data, pattern = "fragments.tsv.gz$")

```


<br><br><br>


To allow comparison of the different conditions, we need to merge the different seurat object. As the peaks were identified independently in each experiment (peak calling on each sample), the peaks are unlikely to be exactly the same. We therefore need to create a common set of peaks across all the datasets to be merged and quantify this peak set in each experiment prior to merging the objects. We decided to use the reduce() function to merge all the intersecting peaks into one. We used the filtered Grange (QC applied and keep only standard chromosomes).
https://stuartlab.org/signac/articles/merging.html    


```{r, Number of peaks before merging}

cat("Number of cells in each condition : \n",
    names(list_seurat_obj[1]), ncol(x = list_seurat_obj[[1]]), "\n",
    names(list_seurat_obj[2]), ncol(x = list_seurat_obj[[2]]), "\n",
    names(list_seurat_obj[3]), ncol(x = list_seurat_obj[[3]]), "\n",
    names(list_seurat_obj[4]), ncol(x = list_seurat_obj[[4]]), "\n",
    "Number of peaks set in each condition : \n",
    names(list_seurat_obj[1]), nrow(x = list_seurat_obj[[1]]), "\n",
    names(list_seurat_obj[2]), nrow(x = list_seurat_obj[[2]]), "\n",
    names(list_seurat_obj[3]), nrow(x = list_seurat_obj[[3]]), "\n",
    names(list_seurat_obj[4]), nrow(x = list_seurat_obj[[4]]), "\n"
)

```

```{r, Merge with common list of peaks}

# Create a unified set of peaks to quantify in each dataset => MANUAL REDUCE
combined.peaks = reduce(x = c(
  list_seurat_obj[[1]]@assays$peaks@ranges,
  list_seurat_obj[[2]]@assays$peaks@ranges,
  list_seurat_obj[[3]]@assays$peaks@ranges,
  list_seurat_obj[[4]]@assays$peaks@ranges
  )) 

# Filter out bad peaks based on length
peakwidths <- width(combined.peaks)
combined.peaks <- combined.peaks[peakwidths  < 10000 & peakwidths > 20]
combined.peaks

list_seurat_combined = list()

for (i in 1:length(list_seurat_obj)) {
  
  frags <- CreateFragmentObject(
    path = paste0(directory_data, 
                  str_subset(string = fragment_list, 
                             pattern = names(list_seurat_obj[i]))),
    cells = colnames(list_seurat_obj[[i]])
    )
  
  matrix = FeatureMatrix(
    fragments = frags,
    features = combined.peaks,
    cells = colnames(list_seurat_obj[[i]])
  )
  
  assay = CreateChromatinAssay(counts = matrix,
                               fragments = frags,
                               genome = "hg38")
  
  seurat = CreateSeuratObject(counts = assay,
                              assay = "peaks", # don"t change the name
                              meta.data = list_seurat_obj[[i]]@meta.data)
  
  seurat$orig.ident = names(list_seurat_obj[i])
    
  list_seurat_combined[[names(list_seurat_obj[i])]] = seurat
  
} # très très long ! 

seurat_common_merged = merge(
  x = list_seurat_combined[[1]],
  y = list_seurat_combined[c(-1)],
  add.cell.ids = names(list_seurat_combined),
  project = "10XscATAC-seq"
)

Idents(object = seurat_common_merged) <- seurat_common_merged$orig.ident

cat("Total number of cells in dataset : ", ncol(seurat_common_merged), "\n",
    "Total number of peaks in common set of features : ", nrow(seurat_common_merged))

```

```{r, Save outputs}

saveRDS(object = seurat_common_merged, 
        file = paste0(directory_output, "merged_seurat_qc_filter.rds"))

```


<br><br><br>


```{r, Rsession}

end_time = Sys.time()
cat("Total execution time : ", as.numeric (end_time - start_time, units = "mins"), "minutes")

# Show package version
sessionInfo()

```

```{r, results="hide"}

# Clean working space and memory 
rm(list = ls())
gc()

```
