---
title: "scATACseq_R-SKMK_ReportCreateQuality"
author: "LaÃ«titia Racine"
date: "2022-09-22"
subtitle: "Last modification : `r format(Sys.time(), '%d %B, %Y')`"
output:
  html_document:
    code_folding: "hide"
    toc: true
    toc_float: false
    theme: journal
---

<style>
body {text-align: justify}
</style>

```{r, Setup, include=F}

knitr::opts_chunk$set(echo = TRUE)
knitr::opts_chunk$set(warning = FALSE)
knitr::opts_chunk$set(message = FALSE)

```

```{r, Dependencies}

library(knitr)
library(Seurat)
library(Signac)
library(stringr)
library(gridExtra)
library(kableExtra)
library(dplyr)

## Problem with CreateChromatinAssay because of GenomeInfoDb
## Update Bioconductor and update GenomeInfoDb package
# if (!require("BiocManager", quietly = TRUE))
#     install.packages("BiocManager")
# BiocManager::install(version = "3.16")
# BiocManager::install("GenomeInfoDb")
# remotes::install_github("Bioconductor/GenomeInfoDb")
## Then restart R session

```

```{r, Working directory }

# Load working directories
directory = str_extract(string = getwd(), pattern = "[:graph:]+(?=bin)")
directory_data = paste0(directory, "data/scATACseq/")
start_time = Sys.time()

# Create a unique folder for output corresponding to the date of the day
current_date = format(Sys.time(), "%Y%m%d")
directory_skmk = paste0(directory, "exp/scATACseq_SKMK_preprocessing/")

```



<br><br><br>



# Overview

This code aims at explaining what was performed in scATACseq_SKMK_CreateSeuratObject.R and scATAC_SKMK_QualityControl.R. Indeed, to perform the first steps of the preprocessing of scATACseq dataset, the command line were adapted to be used with Snakefile in order to parallelize the treatment of the four conditions (24h CTRL, DON, 2DG and AOA).  
  
Based on documentation :      
- https://stuartlab.org/signac/articles/pbmc_vignette.html    
- https://caleblareau.com/cshl-ast2019-scATAC/code/scATAC-tutorial.html        
- https://stuartlab.org/signac/articles/mouse_brain_vignette.html       
- https://support.10xgenomics.com/single-cell-atac/software/pipelines/latest/output/fragments    
- https://stuartlab.org/signac/reference/createchromatinassay  
- https://support.10xgenomics.com/single-cell-atac/software/pipelines/latest/glossary  
- https://support.10xgenomics.com/single-cell-atac/software/pipelines/latest/output/singlecell  
- https://github.com/stuart-lab/signac/issues/826  
- https://support.bioconductor.org/p/9138023/#9138215  
- https://github.com/Bioconductor/GenomeInfoDb/issues/82  
- https://rdrr.io/github/Bioconductor/GenomeInfoDb/  
    
Glossary :  
- Fragment : a piece of genomic DNA, bounded by two adjacent cut sites, that has been converted into a sequencer-compatible molecule with an attached cell-barcode. The alignment interval of the fragment is obtained by correcting the alignment interval of the sequenced fragment by +4 bp on the left end of the fragment, and -5bp on the right end (where left and right are relative to genomic coordinates).  
- Cell Barcode : any barcodes that have been determined by the 'cell-calling' step of the pipeline to be associated with cells.  
- Peak: a compact region of the genome identified as having 'open chromatin' due to an enrichment of cut-sites inside the region. => made of fragments  


    
<br><br><br>  
  

  
# scATACseq_SKMK_CreateSeuratObject - Report

<br>

This code aims at creating a Seurat object with chromatin Assay using the peak/cell matrix, the cell metadata and fragment file generated by CellRanger software.  

<br>

## Script command lines

```{r, eval = FALSE}

#**************
# Input loading
#**************

matrixh5 = arguments$matrixh5_file
fragment = arguments$fragment_file
singlecell = arguments$metadata_file

name_cond = str_extract(string = singlecell, pattern = "[:alnum:]+(?=_singlecell.csv)")

```

<br>

The *matrix.h5* file is a matrix with one **peak** per row (chr:start-end) and one **cell barcode** per column. Each value in the matrix represents the number of Tn5 integration sites for each single barcode (i.e. a cell) that map within each peak (i.e. number of fragments that mapped).  
The *fragment* file contains a full list of all unique fragments across single cells even fragments that didn't map to peaks. It's a table with a row per fragment and as columns : chr, start, end, barcode of the cell, read pairs associated with this fragment (this includes the read pair marked unique and all duplicate read pairs).      
The *singlecell* file is a table which provides QC information associated with the fragments per barcode. We can find inside the barcodes kept in the dataset (barcodes from the matrix) and other barcodes that were excluded.   

<br>

```{r, eval = FALSE}

#*********************
# Create Seurat Object
#*********************

counts <- Seurat::Read10X_h5(filename = matrixh5)

metadata <- read.csv(
  file = singlecell,
  header = TRUE,
  row.names = 1
)

chrom_assay <- CreateChromatinAssay(
  counts = counts,
  sep = c(":", "-"),
  genome = 'hg38', 
  fragments = fragment,
  min.cells = 100,
  min.features = 200  # features = peak
)

seurat_obj <- CreateSeuratObject(
  counts = chrom_assay,
  assay = "peaks",
  meta.data = metadata)

```

<br>

The *ChromatinAssay* is created with the fragment file and the matrix.h5. The genome used for mapping (hg38) is indicated to appear in metadata. Some important parameters that perform **filters on raw data** appear here :    
- min.cells = 100 : Include features detected in at least this many cells. Will subset the counts matrix as well.    
- min.features = 200 : Include cells where at least this many features are detected.      
Then, this ChromatinAssay is stored in a Seurat Assay and metadata for the cells are added.   

<br>

```{r, eval = FALSE}

#*****************
# Add annotations
#*****************

# Extract annotations from EnsDb database
annotations <- GetGRangesFromEnsDb(ensdb = EnsDb.Hsapiens.v86, standard.chromosomes = TRUE)
seqlevelsStyle(annotations) = "UCSC"
print(table(annotations$type))

# Add annotation database into seurat object (needed for downstream analysis (QC functions))
Annotation(seurat_obj) = annotations

# Visualize first object of the list to see the annotations
seurat_obj@assays$peaks@annotation


#*****************
# Update orig.ident
#*****************

Idents(object = seurat_obj) = name_cond
seurat_obj$orig.ident = name_cond

#*****************
# Add peak_name as grange metadata
#*****************

peaks_name = data.frame(paste0(seurat_obj@assays$peaks@ranges@seqnames, "-", seurat_obj@assays$peaks@ranges@ranges))
colnames(peaks_name) = "peaks_name"
mcols(seurat_obj@assays$peaks@ranges)$peaks_name = peaks_name

#************
# Save output
#************

saveRDS(object = seurat_obj, file = arguments$output)

```

<br>

Gene annotation for the human genome (from database EnsDb.Hsapiens.v86 for hg38) is added. This can serve for downstream functions to pull the gene annotation information directly from the object. Then, the name of the condition (CTRL/DON/2DG/AOA) is added as metadata in orig.ident.     

<br>

## Visualization of created objects

<br>

**Extract initial matrix.h5**
```{r, eval = TRUE, results='hold'}

matrix_list = list.files(path = directory_data, pattern = ".h5")
matrix_list = paste0(directory_data, matrix_list)
cond_name = lapply(matrix_list, function(x) {str_extract(string = x, pattern = "(?<=scATACseq/)[:alnum:]+(?=_)")})
cond_name = unlist(cond_name)
matrix_list = lapply(matrix_list, function(x) {Seurat::Read10X_h5(x)})
names(matrix_list) = cond_name

tab_mat = full_join(x = data.frame(unlist(lapply(matrix_list, nrow))) %>% tibble::rownames_to_column("condition"),
                y = data.frame(unlist(lapply(matrix_list, ncol))) %>% tibble::rownames_to_column("condition"),
                by = "condition")
colnames(tab_mat) = c("condition", "Nbpeaks", "Nbcells")

tab_mat %>%
  kable(caption = "Initial matrices of counts (from Cell Ranger output)") %>%
  kable_styling()

```

Those matrices contain the same list of peaks that we can find in the .bed files from Cell Ranger output folder.  

<br>

**Compare with matrix count in the Seurat objects created**
```{r, eval = TRUE, results='hold'}

seurat_list = list.files(path = directory_skmk, pattern = "_seurat.rds")
seurat_list = paste0(directory_skmk, seurat_list)
cond_name = lapply(seurat_list, function(x) {str_extract(string = x, pattern = "(?<=SKMK_preprocessing/)[:alnum:]+(?=_)")})
cond_name = unlist(cond_name)
seurat_list = lapply(seurat_list, function(x) {readRDS(x)})
names(seurat_list) = cond_name

tab_seurat = full_join(x = data.frame(unlist(lapply(seurat_list, nrow))) %>% tibble::rownames_to_column("condition"),
                y = data.frame(unlist(lapply(seurat_list, ncol))) %>% tibble::rownames_to_column("condition"),
                by = "condition")
colnames(tab_seurat) = c("condition", "Nbpeaks", "Nbcells")

tab_seurat %>%
  kable(caption = "Matrices of counts in ChromatinAssay from Seurat object") %>%
  kable_styling()

```

Some peaks and cells were removed during the creation of the seurat object because of the two filters applied on data.  

<br>

**Visualize one Seurat Object structure (Example with CTRL condition)**
```{r, eval = TRUE}

# Load file 
CTRL_seurat_obj = seurat_list[["CTRL"]]
# Global structure
CTRL_seurat_obj
## chromatin assay with the peaks
CTRL_seurat_obj@assays$peaks
## list of peaks from chromatin assay
CTRL_seurat_obj@assays$peaks@ranges
## raw count matrix store in counts.slot (first 10 rows and column)
as.data.frame(CTRL_seurat_obj@assays$peaks@counts)[1:10, 1:10]
## normalized count matrix store in data.slot (first 10 rows and column)
as.data.frame(CTRL_seurat_obj@assays$peaks@data)[1:10, 1:10]
# => same as the raw matrix because the normalization is not done yet

# database annotations
CTRL_seurat_obj@assays$peaks@annotation

```



<br><br><br>



# scATACseq_SKMK_QualityControl - Report

This second code aims at computing some QC metrics to assess data quality and to filter peaks and cells that have bad quality (outliers).  
We try to find the most fitted limits for all conditions to apply same filtering on all conditions.    
  
## Script command lines

<br>

```{r, eval = FALSE}

#**************
# Input loading
#**************

seurat_obj = readRDS(file = arguments$seurat_obj)
name_cond = arguments$cond

```

<br>

We work from the output of the previous script : the seurat object annotated with Chromatin Assay. This object already passed two filters on cells and peaks. 

<br>

```{r, eval = FALSE}

#************************
# Initial number of cells
#************************

# Extract number of cells and peaks 
nbcells = ncol(seurat_obj)
nbpeaks = nrow(seurat_obj)

# Store data in a tab
df_summary = data.frame("Conditions" = name_cond,
                        "Initial_Cell_Number" = nbcells,
                        "Initial_Peak_Number" = nbpeaks)

```

<br>

**Nucleosome banding pattern**: it represents fragment length periodicity.  
The histogram of DNA fragment sizes should exhibit a strong nucleosome banding pattern corresponding to the length of DNA wrapped around a single nucleosome.     

<br>

```{r, eval = FALSE}

#***************************
# Nucleosome banding pattern
#***************************

seurat_obj = NucleosomeSignal(object = seurat_obj)
seurat_obj$nucleosome_group = ifelse(seurat_obj$nucleosome_signal > "2", 'NS > 2', 'NS < 2')
  
qc_plot_nucl_fragments = FragmentHistogram(object = seurat_obj) +
  geom_vline(xintercept = 147, color = "red") +  # no nucleosome limit
  geom_vline(xintercept = 294, color = "red") +  # one nucleosome limit
  geom_vline(xintercept = 441, color = "red") +  # two nucleosomes limit
  geom_vline(xintercept = 588, color = "red") +  # three nucleosomes limit
  labs(title = "Nucleosome banding pattern") +
  theme(title = element_text(face = "bold"))
  
qc_plot_nucl_signal = VlnPlot(
  object = seurat_obj,
  features = 'nucleosome_signal',
  pt.size = 0.1) + 
  theme(legend.position = 'none',
        axis.title.x = element_blank(),
        axis.text.x = element_blank()) +  
  geom_hline(yintercept = 2, color = "red") 
  
qc_plot_nucl_quality = FragmentHistogram(
    object = seurat_obj, 
    group.by = 'nucleosome_group') +
    labs(title = "Nucleosome signal") +
    theme(title = element_text(face = "bold"))

```

<br>

**Fragments number and blacklist**    
- Ratio reads in genomic blacklist regions : cells with a high proportion of reads mapping to these areas often represent technical artifacts and should be removed.    
- Fraction of fragments in peaks: represents the fraction of all fragments that fall within ATAC-seq peaks. Cells with low values (i.e. <15-20%) often represent low-quality cells or technical artifacts that should be removed.  

<br>

```{r, eval = FALSE}

#**********************************
# Number of fragments and blacklist
#**********************************

seurat_obj$pct_reads_in_peaks = seurat_obj$peak_region_fragments / 
    seurat_obj$passed_filters * 100
seurat_obj$blacklist_ratio = seurat_obj$blacklist_region_fragments /
    seurat_obj$peak_region_fragments
  
qc_plot_pct_reads_vln = VlnPlot(object = seurat_obj,
                                features = 'pct_reads_in_peaks',
                                pt.size = 0.1) + 
    theme(legend.position = 'none',
          axis.title.x = element_blank(),
          axis.text.x = element_blank())
  
qc_plot_pct_reads_point = qplot(log10(seurat_obj$passed_filters),
                                  seurat_obj$pct_reads_in_peaks) +
    labs(x = "log10 # of fragments", y = "% reads in peaks") +
    geom_hline(yintercept = 60, color = "firebrick") +
    geom_vline(xintercept = log10(5000), color = "firebrick")
  
qc_plot_blacklist = VlnPlot(object = seurat_obj,
                            features = 'blacklist_ratio',
                            pt.size = 0.1) + 
    theme(legend.position = 'none',
          axis.title.x = element_blank(),
          axis.text.x = element_blank())
  
qc_plot_peak_region_fragments = VlnPlot(object = seurat_obj,
                                        features = 'peak_region_fragments',
                                        pt.size = 0.1) + 
    theme(legend.position = 'none',
          axis.title.x = element_blank(),
          axis.text.x = element_blank())

```

<br>

**Transcriptional start site enrichment score** : Poor ATAC-seq experiments typically will have a low TSS enrichment score.     

<br>

```{r, eval = FALSE}

#*********************
# TSS enrichment score 
#*********************

seurat_obj = TSSEnrichment(object = seurat_obj, fast = FALSE)
seurat_obj$high.tss = ifelse(seurat_obj$TSS.enrichment > 2, 'High', 'Low')
seurat_obj$pct_reads_in_TSS = seurat_obj$TSS_fragments / seurat_obj$passed_filters * 100
  
qc_plot_tss = TSSPlot(seurat_obj, group.by = 'high.tss') + 
    NoLegend() +
    labs(title = "TSS enrichment score") +
    theme(title = element_text(face = "bold"))
  
qc_plot_tss_reads =  qplot(seurat_obj$pct_reads_in_TSS, seurat_obj$pct_reads_in_peaks) +
    labs(x = "% reads in TSS", y = "% reads in peaks") +
    geom_abline(intercept = 0, slope = 1, linetype = 2, color = "firebrick")

```

```{r, eval = FALSE}

#*************************************
# Visualize QC metrics and save plots
#*************************************

list_graphs = list(qc_plot_nucl_fragments, qc_plot_nucl_signal, qc_plot_nucl_quality,
              qc_plot_pct_reads_vln, qc_plot_pct_reads_point, qc_plot_blacklist,
              qc_plot_peak_region_fragments, qc_plot_tss, qc_plot_tss_reads)
ggsave(plot = arrangeGrob(grobs = list_graphs, 
                          ncol = 3, nrow = 3,
                          top = textGrob(paste("QC metrics for", name_cond))),
         filename = arguments$output_plots_panel,
         width = 40, height = 40)

saveRDS(object = list_graphs, file = arguments$output_plots_list)
saveRDS(object = seurat_obj, file = arguments$output_seurat)

```


<br>

We filter the cells based on those parameters :      
- nucleosome_signal < 2      
- peak_region_fragments > 3000 & peak_region_fragments < 50000    
- pct_reads_in_peaks > 50      
- blacklist_ratio < 0.05      
- TSS.enrichment > 2      
- keep only standard chromosomes        
NB : nucleosome signal limit at 2, possible to adapt depending the data        
  
The cells are filtered out from the count matrix with subset() function.  
The peaks are removed from the Grange with tidyChromosomes() function.  

<br>


```{r, eval = FALSE}

# Initialize summary df with initial numbers of cells and peaks
vec_cell = c(paste0(name_cond, "_NbCells"))
vec_peak = c(paste0(name_cond, "_NbPeaks"))
df_filter = data.frame(
  "Condition" = c(vec_cell, vec_peak),
  "CreateChromatinAssay_filter" = c(df_summary$Initial_Cell_Number, 
                                    df_summary$Initial_Peak_Number))

#***********************************
# Number of cells after QC filtering  
#***********************************

# Apply first cell filter on nucleosome signal 
seurat_obj = subset(x = seurat_obj, subset = nucleosome_signal < 2)
df_filter$Nucleosome_filter = c(ncol(seurat_obj), nrow(seurat_obj))

# Apply cell filter on number of fragments in cell
seurat_obj =  subset(x = seurat_obj, 
                     subset = peak_region_fragments > 3000 & 
                       peak_region_fragments < 50000)
df_filter$Fragments_filter =  c(ncol(seurat_obj), nrow(seurat_obj))

# Apply cell filter on percentage of fragments in peaks
seurat_obj =  subset(x = seurat_obj, subset = pct_reads_in_peaks > 50)
df_filter$PctReads_filter =  c(ncol(seurat_obj), nrow(seurat_obj))

# Apply cell filter on blacklist fragments
seurat_obj =  subset(x = seurat_obj, subset = blacklist_ratio < 0.05)
df_filter$Blacklist_filter =  c(ncol(seurat_obj), nrow(seurat_obj))

# Apply cell filter on TSS enrichment score
seurat_obj =  subset(x = seurat_obj, subset = TSS.enrichment > 2)
df_filter$TSS_filter =  c(ncol(seurat_obj), nrow(seurat_obj))

#***********************************
# Number of peaks after QC filtering  
#***********************************

## Clean chromosome list - Keep only standard chromosomes
# Filter in the Grange object
seurat_obj@assays$peaks@ranges = tidyChromosomes(gr = seurat_obj@assays$peaks@ranges,
                                                 keep.X = TRUE,
                                                 keep.Y = TRUE,
                                                 keep.M = FALSE,
                                                 keep.nonstandard = FALSE,
                                                 genome = "hg38")
# Subset in the Count matrix as well
list_peak_clean = unlist(as.vector(seurat_obj@assays$peaks@ranges$peaks_name))
table(rownames(seurat_obj) %in% list_peak_clean) # we need to remove the FALSE peaks
seurat_obj = subset(seurat_obj, features = list_peak_clean) # works only because we have only one assay
table(rownames(seurat_obj) %in% list_peak_clean)
df_filter$Chromosome_filter = c(ncol(seurat_obj), nrow(seurat_obj))

```

<br>

As a reminder :    
- *Grange* = peaks list with their genomic location (chr, start, end)  
- *Count* = count matrix with list of peaks and number of fragments per cell for each peak  
Note : changing grange by subsetting doesn't affect the count matrix. It's important to manually filter the count matrix to keep the same list of peaks in both elements.
https://www.biostars.org/p/9466838/  
https://kasperdanielhansen.github.io/genbioconductor/html/GenomicRanges_seqinfo.html  
https://web.mit.edu/~r/current/arch/i386_linux26/lib/R/library/GenomeInfoDb/html/seqinfo.html  

<br>

```{r, eval = FALSE}

#**********************
# Save filtered output
#**********************

saveRDS(object = seurat_obj, file = arguments$output_seurat_filtered)
write.table(x = df_filter, file = arguments$output_tab, row.names = FALSE)

```

<br>


## Filtering summary for all conditions

<br>

```{r}

# Load summary tab for each condition
CTRL_df_filter = read.csv2(file = paste0(directory_skmk, "CTRL_df_filter.csv"), sep = "")
DDG_df_filter = read.csv2(file = paste0(directory_skmk, "2DG_df_filter.csv"), sep = "")
DON_df_filter = read.csv2(file = paste0(directory_skmk, "DON_df_filter.csv"), sep = "")
AOA_df_filter = read.csv2(file = paste0(directory_skmk, "AOA_df_filter.csv"), sep = "")

# Merge tab into one global tab
df_filter_glob = rbind(CTRL_df_filter, DDG_df_filter, DON_df_filter, AOA_df_filter)

df_filter_glob %>%
  kable() %>%
  kable_styling()

```



<br><br><br>



# Conclusion

<br>

At the beginning, we start with the list of peaks and cells that are included in the matrix.h5. Those lists of peaks are the same as the one we can find in the .bed files.      
Then, those information are converted into Seurat Object and the peaks/cells are filtered based on QC values.    
=> initial count matrix       
=> filter 1 on cells : Keep cells in which at least 200 peaks were detected     
=> filter 2 on peaks : Keep peaks detected in at least 100 cells  
=> filter 3 on peaks : Nucleosome signal < 2    
=> filter 4 on cells : Number of fragments in cell > 3000 and < 50000    
=> filter 5 on cells : Percentage of fragment in peaks > 50  
=> filter 6 on cells : Blacklist ratio < 0.05   
=> filter 7 on cells : TSS score > 2    
=> filter 8 on peaks : Keep only standard chromosomes    
   
```{r, Save new list of peaks in .bed format}

write.csv2(x = data.frame(seurat_list[["CTRL"]]@assays$peaks@ranges), 
           file = paste0(directory_skmk, "CTRL_peaks_filtered.csv"),
           row.names = FALSE)

write.csv2(x = data.frame(seurat_list[["DON"]]@assays$peaks@ranges), 
           file = paste0(directory_skmk, "DON_peaks_filtered.csv"),
           row.names = FALSE)

write.csv2(x = data.frame(seurat_list[["2DG"]]@assays$peaks@ranges),
           file = paste0(directory_skmk, "2DG_peaks_filtered.csv"),
           row.names = FALSE)

write.csv2(x = data.frame(seurat_list[["AOA"]]@assays$peaks@ranges), 
           file = paste0(directory_skmk, "AOA_peaks_filtered.csv"),
           row.names = FALSE)

```



```{r, Rsession}

end_time = Sys.time()
cat("Total execution time : ", as.numeric (end_time - start_time, units = "mins"), "minutes")

# Show package version
sessionInfo()

```

```{r, results="hide"}

# Clean working space and memory 
rm(list = ls())
gc()

```

