---
title: "scATACseq_R_AddAnnotations"
author: "LaÃ«titia Racine"
date: "2022-09-16"
subtitle: "Last modification : `r format(Sys.time(), '%d %B, %Y')`"
output:
  html_document:
    code_folding: "hide"
    toc: false
    toc_float: false
    theme: journal
---

<style>
body {text-align: justify}
</style>

```{r, Setup, include=FALSE}

knitr::opts_chunk$set(echo = TRUE)
knitr::opts_chunk$set(warning = FALSE)
knitr::opts_chunk$set(message = FALSE)

```

```{r, Dependencies}

library(EnsDb.Hsapiens.v86)   # v.86 = hg38 | v.75 = hg19
library(GenomeInfoDb)
library(knitr)
library(stringr)
library(Signac)

```

```{r, Working directories and external script}

directory = str_extract(string = getwd(), pattern = "[:graph:]+(?=bin)")
start_time = Sys.time()

# Create a unique folder for output corresponding to the date of the day
current_date = format(Sys.time(), "%Y%m%d")
dir.create(path = paste0(directory,"exp/","scATACseq_R_AddAnnotations/"))
dir.create(path = paste0(directory,"exp/","scATACseq_R_AddAnnotations/", current_date))
directory_output = paste0(directory,"exp/","scATACseq_R_AddAnnotations/", current_date, "/")

# Load external script with functions and constants
source(file = paste0(directory, "bin/", "functions_constants.R"))

```

```{r, Input loading}

# Load seurat objects 
dir_seurat = pic_last_dir(paste0(directory, "exp/", "scATACseq_R_MergeObjectsCommonPeaks/"))
seurat_common_merged = readRDS(paste0(dir_seurat, "/", "merged_seurat_qc_filter.rds"))

# Load grange with hg38 annotations
dir_annot = pic_last_dir(paste0(directory, "exp/", "AnnotationsFiles_createGrange/"))
hg38_annotations = readRDS(paste0(dir_annot, "/", "hg38_annotations_gr.rds"))

```


<br><br><br>

**Add annotations database in the annotations slot**
```{r, Add annotations database}

# Extract annotations from EnsDb database
annotations <- GetGRangesFromEnsDb(ensdb = EnsDb.Hsapiens.v86, standard.chromosomes = TRUE)
seqlevelsStyle(annotations) = "UCSC"
print(table(annotations$type))

# Add annotation database into seurat object 
Annotation(seurat_common_merged) = annotations

```

<br>

**Add peak_name in the metadata**
```{r, Add peak_name metadata}

add_peak_name = function(seurat_obj) {
  print(seurat_obj@assays$peaks@ranges)
  names_peaks = paste0(seurat_obj@assays$peaks@ranges@seqnames, "-", seurat_obj@assays$peaks@ranges@ranges)
  mcols(seurat_obj@assays$peaks@ranges)$peak_name = names_peaks
  print(seurat_obj@assays$peaks@ranges)
  return(seurat_obj)
}

seurat_common_merged = add_peak_name(seurat_common_merged)

```

<br>

**Add peaks annotation in the metadata**
```{r, Add peaks annotations}

add_peak_annot = function(seurat_obj, annotations_gr) {
  
  # First a matrix is created filled with FALSE and added to the Grange
  annotations_types = levels(factor(annotations_gr$type))
  metadata = matrix(FALSE, ncol = length(annotations_types), 
                    nrow = length(seurat_obj@assays$peaks@ranges))
  colnames(metadata) = annotations_types
  mcols(seurat_obj@assays$peaks@ranges) = append(mcols(seurat_obj@assays$peaks@ranges), metadata) 
  
  # For each of the annotations types an overlap is calculated 
  # and used to assigned the peak as TRUE when overlapping with the annotation
  for (i in 1:ncol(metadata)){
    sub_annot = annotations_gr[annotations_gr$type == annotations_types[i]]
    overlaps = findOverlaps(seurat_obj@assays$peaks@ranges, sub_annot)
    mcols(seurat_obj@assays$peaks@ranges)[queryHits(overlaps),i] = TRUE
  }
  
  # Return output
  return(seurat_obj)
  
}

seurat_common_merged = add_peak_annot(seurat_common_merged, hg38_annotations)
print(seurat_common_merged@assays$peaks@ranges)

```

<br>

```{r, Save outputs}

saveRDS(seurat_common_merged, paste0(directory_output, "merged_seurat_qc_filter_annot.rds"))

```


<br><br><br>


```{r, Rsession}

end_time = Sys.time()
cat("Total execution time : ", as.numeric (end_time - start_time, units = "mins"), "minutes")

# Show package version
sessionInfo()

```

```{r, results="hide"}

# Clean working space and memory 
rm(list = ls())
gc()


```

