---
title: "scRNAseq_CITEseq_R_DWNS_CompareConditions_GO"
author: "LaÃ«titia Racine"
date: "2022-12-15"
subtitle: "Last modification : `r format(Sys.time(), '%d %B, %Y')`"
output:
  html_document:
    code_folding: "hide"
    toc: true
    toc_float: false
    theme: journal
---

<style>
body {text-align: justify}
</style>

```{r setup, include=FALSE}

knitr::opts_chunk$set(echo = TRUE)
knitr::opts_chunk$set(warning = FALSE)
knitr::opts_chunk$set(message = FALSE)

```

```{r, Dependencies}

library(Seurat)
library(dplyr)
library(stringr)
library(ggplot2)
library(gridExtra)
library(tidyverse)
library(ggrepel)
library(org.Hs.eg.db) # GO functions
library(clusterProfiler) # GO functions
library(kableExtra)

# https://support.bioconductor.org/p/38541/
library(foreach)
library(doParallel)
library(parallel)

```

# Overview

https://ycl6.github.io/GO-Enrichment-Analysis-Demo/

```{r, Working directories and external script}

# Load working directories
directory = str_extract(string = getwd(), pattern = "[:graph:]+(?=bin)")
start_time = Sys.time()

# Create a unique folder for output corresponding to the date of the day
current_date = format(Sys.time(), "%Y%m%d")
dir.create(path = paste0(directory, "exp/scRNAseq_CITEseq_R_DWNS_CompareConditions/"))
dir.create(path = paste0(directory, "exp/scRNAseq_CITEseq_R_DWNS_CompareConditions/", current_date))
directory_output = paste0(directory, "exp/scRNAseq_CITEseq_R_DWNS_CompareConditions/", current_date, "/")

# Load external script with functions and constants
source(file = paste0(directory, "bin/", "functions_constants.R"))

```

```{r, Input loading}

dir = pic_last_dir(paste0(directory, "exp/", "scRNAseq_CITEseq_R_ReductionDimension/"))
merged_all = readRDS(file = paste0(dir, "/", "merged_all_1-40dim.rds"))

dir_corr = pic_last_dir(paste0(directory, "exp/", "scRNAseq_CITEseq_R_GenomePosition/"))
tab_corr = read.csv2(paste0(dir_corr, "/scRNAseq_correspondence_tab_EnsemblID_geneName_genomePos.csv")) %>% 
    dplyr::select(gene_symbol_dupl, ensembl_gene_id) %>%
    dplyr::rename(gene = "gene_symbol_dupl")

gene_universe = read.delim(paste0(directory, "data/scRNAseq/", 
                                  "scRNAseq_CTRL_sample_feature_bc_matrix_features.tsv.gz"), 
                                  header = FALSE)[,1]

```

Choose the combinations we want to analyse.
```{r, Combinations}

combination = list(c("CTRL","AOA"), c("CTRL","2DG"), c("CTRL","DON"),
                   c("CTRL", "VPA"), c("CTRL","CTRLaK"), c("2DG","2DGaK"),
                   c("DON","DONaK"), c("AOA","AOAaK"), c("CTRLaK","2DGaK"),
                   c("CTRLaK","AOAaK"), c("CTRLaK","DONaK"))

```



# Qualitative analysis : expressed / no expressed

## Binary matrix 

On/Off analysis : we created a binary matrix expressed/no expressed to remove the variation of the number of UMI. 

```{r, Binary matrix expressed/no expressed}

# extract count matrix from SCT assay and transform into data frame
# we can work with counts slot because it's not a quantitative analysis 
# just expressed/no expressed, no need for normalized data
# Split into list of expressed and no expressed genes
gene_expr_fun = function(seurat_obj, condition) {
  
  tab = subset(seurat_obj, subset = orig.ident %in% condition)
  tab = as.data.frame(tab@assays$SCT@counts) %>% mutate_all(as.logical)
  tab = data.frame(gene = rownames(tab), expression = rowSums(tab))
  tab$expression = as.logical(tab$expression)
  tab_exp = tab %>% dplyr::filter(expression == TRUE)
  tab_noexp = tab %>% dplyr::filter(expression == FALSE)
  return(list(tab_exp = tab_exp, tab_noexp = tab_noexp))
  
}

list_bin_mat = list("CTRL" = gene_expr_fun(merged_all, c("CTRL")),
                    "AOA" = gene_expr_fun(merged_all, "AOA"),
                    "2DG" = gene_expr_fun(merged_all, "2DG"),
                    "DON" = gene_expr_fun(merged_all, "DON"),
                    "CTRLaK" = gene_expr_fun(merged_all, "CTRLaK"),
                    "AOAaK" = gene_expr_fun(merged_all, "AOAaK"),
                    "2DGaK" = gene_expr_fun(merged_all, "2DGaK"),
                    "DONaK" = gene_expr_fun(merged_all, "DONaK"),
                    "VPA" = gene_expr_fun(merged_all, "VPA"))

saveRDS(list_bin_mat, paste0(directory_output, "list_binary_mat.rds"))

```

## Extract list of genes 

```{r}

# List to be completed 
list_genes = list()

# Loop for combination of comparisons
for (i in 1:length(combination)) {
  
  cond = paste(combination[[i]], collapse = "_")
  name_cond1 = combination[[i]][1]
  name_cond2 = combination[[i]][2]
  expr_cond1 = list_bin_mat[[name_cond1]]$tab_exp
  expr_cond2 = list_bin_mat[[name_cond2]]$tab_exp
  
  #################################
  ### Genes lost across conditions
  #################################
  
  # elements in cond1 but not in cond2
  diff_cond1_cond2 = setdiff(expr_cond1$gene, expr_cond2$gene)
  # elements in cond2 but not in cond1
  diff_cond2_cond1 = setdiff(expr_cond2$gene, expr_cond1$gene)
  # save in the list 
  list_genes[[paste0("diff_", name_cond1, "_", name_cond2)]] = diff_cond1_cond2
  list_genes[[paste0("diff_", name_cond2, "_", name_cond1)]] = diff_cond2_cond1
  
  #####################################
  ### Genes conserved across conditions
  #####################################
  
  # common elements between cond1 and cond2
  inter_cond1_cond2 = intersect(expr_cond1$gene, expr_cond2$gene)
  # save in the list
  list_genes[[paste0("inter_", name_cond1, "_", name_cond2)]] = inter_cond1_cond2
  
}

# Summarize in a dataframe
tab_diff = data.frame(comparison = names(list_genes),
                      nbgenes_comparison = lengths(list_genes)) %>%
  tidyr::separate(col = comparison, into = c("comparison", "cond1", "cond2"), sep = "_", remove = TRUE) %>%
  dplyr::select(cond1, cond2, comparison, nbgenes_comparison)

tab_genes_expressed = as.data.frame(sapply(unlist(list_bin_mat, recursive=FALSE), nrow)) %>%
  tibble::rownames_to_column("condition") %>%
  dplyr::filter(str_detect(condition, "tab_exp")) %>%
  tidyr::separate(col = condition, into = c("condition", "trash"), remove = TRUE)
colnames(tab_genes_expressed) = c("condition", "trash", "nbgenes_expressed")
tab_genes_expressed = tab_genes_expressed %>% dplyr::select(-trash)

tab_diff = tab_diff %>% dplyr::rename(condition = "cond1")
tab_diff = full_join(tab_diff, tab_genes_expressed, by = "condition") %>%
  dplyr::rename(cond1 = "condition", nbgenes_expressed_cond1 = "nbgenes_expressed") %>%
  dplyr::rename(condition = "cond2")
tab_diff = full_join(tab_diff, tab_genes_expressed, by = "condition") %>%
  dplyr::rename(cond2 = "condition", nbgenes_expressed_cond2 = "nbgenes_expressed") %>%
  dplyr::select(cond1, nbgenes_expressed_cond1, cond2, nbgenes_expressed_cond2, comparison, nbgenes_comparison)

tab_diff %>%
  kable() %>%
  kable_styling()

```

```{r}

#####################################################################
### Genes expressed or not expressed in all conditions simultaneously
#####################################################################
  
all_cond = unlist(list_bin_mat,recursive=FALSE)

all_cond_noexpr = all_cond[str_detect(names(all_cond), "tab_noexp")]
all_cond_noexpr = unlist(all_cond_noexpr, recursive=FALSE)
all_cond_noexpr = all_cond_noexpr[str_detect(names(all_cond_noexpr), "gene")]
genes_absent = Reduce(intersect, all_cond_noexpr)
list_genes[["Never_Expressed"]] = genes_absent

all_cond_expr = all_cond[str_detect(names(all_cond), "tab_exp")]
all_cond_expr = unlist(all_cond_expr, recursive=FALSE)
all_cond_expr = all_cond_expr[str_detect(names(all_cond_expr), "gene")]
genes_all = Reduce(intersect, all_cond_expr)
list_genes[["Always_Expressed"]] = genes_all

cat("Number of genes absent in all conditions (even CTRL) :", length(genes_absent), "\n",
    "Number of genes present in all conditions (even CTRL) :", length(genes_all), "\n")


###########################################################
### Genes expressed in CTRL but not in any other conditions 
###########################################################

# Genes expressed in drug conditions
others = unlist(list_bin_mat[names(list_bin_mat) != "CTRL"], recursive=FALSE)
others = unlist(others, recursive=FALSE)
others = others[str_detect(names(others), "tab_exp.gene")]
others = unique(unlist(others, recursive=FALSE))

# Genes expressed in ctrl condition
ctrl = list_bin_mat[["CTRL"]]$tab_exp$gene

# Genes expressed in ctrl but not in others
diff_ctrl = setdiff(ctrl, others)
list_genes[["CTRL_only"]] = diff_ctrl

cat("Number of genes expressed in CTRL but absent in every other conditions :", 
    length(diff_ctrl), "\n")

###########################################################
### Genes expressed in AOA but not in any other conditions
###########################################################

# Genes expressed in AOA condition
aoa = list_bin_mat[["AOA"]]$tab_exp$gene

# Genes expressed in others conditions
others = unlist(list_bin_mat[names(list_bin_mat) != "AOA"], recursive=FALSE)
others = unlist(others, recursive=FALSE)
others = others[str_detect(names(others), "tab_exp.gene")]
others = unique(unlist(others, recursive=FALSE))

# Genes expressed in ctrl but not in others
diff_aoa = setdiff(aoa, others)
list_genes[["AOA_only"]] = diff_aoa

cat("Number of genes expressed in AOA but absent in every other conditions :", 
    length(diff_aoa), "\n")

###########################################################
### Genes expressed in 2DG but not in any other conditions
###########################################################

# Genes expressed in 2DG condition
ddg = list_bin_mat[["2DG"]]$tab_exp$gene

# Genes expressed in others conditions
others = unlist(list_bin_mat[names(list_bin_mat) != "2DG"], recursive=FALSE)
others = unlist(others, recursive=FALSE)
others = others[str_detect(names(others), "tab_exp.gene")]
others = unique(unlist(others, recursive=FALSE))

# Genes expressed in ctrl but not in others
diff_ddg = setdiff(ddg, others)
list_genes[["2DG_only"]] = diff_ddg

cat("Number of genes expressed in 2DG but absent in every other conditions :", 
    length(diff_ddg), "\n")


```

```{r}

as.data.frame(lengths(list_genes)) %>%
  kable() %>%
  kable_styling()

rm(list=setdiff(ls(), c("list_genes", "tab_corr", "merged_all", 
                        "directory_output", "combination", "list_bin_mat", 
                        "color_code", "current_date", "start_time", 
                        "GO_clusterProfiler_fun", "gene_universe")))

```

## Perform GO on list of genes

```{r, fig.height = 10, fig.width = 10}

list_GO_ego = list()
list_GO_plot = list()

for (i in 1:length(list_genes)) {
  
  print(names(list_genes[i]))
  tab_gene = data.frame(list_genes[i])

  if(nrow(tab_gene) == 0) { 
    print("No gene in the list.") 
    
  } else {
    GO_results = GO_clusterProfiler_fun(tab_corr = tab_corr,
                                        list_gene = tab_gene,
                                        directory_output = directory_output,
                                        gene_universe = gene_universe)
    list_GO_ego[[names(list_genes[i])]] = GO_results
  }
  
} 

list_GO_plot = unlist(list_GO_ego,recursive=FALSE)
list_GO_plot = list_GO_plot[str_detect(names(list_GO_plot), "plot")]
list_GO_plot

```

```{r, Save outputs, fig.show='hide'}

# Save GO list of genes 
saveRDS(list_GO_ego, paste0(directory_output, "list_GO_ego_plot.rds"))

# Save GO plots
for (i in 1:length(list_GO_plot)) {
  plot = list_GO_plot[[i]]
  if (str_detect(names(list_GO_plot[i]), "barplot")) { suffix = "barplot.svg" }
  if (str_detect(names(list_GO_plot[i]), "dotplot")) { suffix = "dotplot.svg" }
  prefix = str_extract(names(list_GO_plot[i]), "[:graph:]+(?=.[:alpha:]{3}plot_25)")
  ggsave(plot = plot,
         filename = paste0(directory_output, prefix, "_GO_plots_", suffix),
         width = 8, height = 12)
}

# Save GO summary for ego
list_GO_summary = unlist(list_GO_ego,recursive=FALSE)
list_GO_summary = list_GO_summary[str_detect(names(list_GO_summary), "ego")]
for (i in 1:length(list_GO_summary)) {
  suffix = str_extract(names(list_GO_summary[i]), "[:graph:]+(?=.ego_result)")
  write.csv2(list_GO_summary[[i]], row.names = FALSE,
             file = paste0(directory_output, "summary_ego_",suffix,".csv")) 
       
}

```

```{r, Rsession}

end_time = Sys.time()
cat("Total execution time : ", as.numeric (end_time - start_time, units = "mins"), "minutes")

# Show package version
sessionInfo()

```

```{r, results = "hide"}


# Clean working space and memory
rm(list = ls())
gc()


```

