---
title: "scRNAseq_CITEseq_R_CreateSeuratObject"
author: "LaÃ«titia Racine"
date: "2021-12-06"
subtitle: "Last modification : `r format(Sys.time(), '%d %B, %Y')`"
output:
  html_document:
    code_folding: "hide"
    toc: true
    toc_float: false
    theme: journal
---

<style>
body {text-align: justify}
</style>

```{r Setup, include=FALSE}

knitr::opts_chunk$set(echo = TRUE)
knitr::opts_chunk$set(warning = FALSE)
knitr::opts_chunk$set(message = FALSE)

```

```{r, Dependencies}

library(Seurat)
library(dplyr)
library(stringr)
library(knitr)

```

<br> 

This code aims to :  
- create count matrix from CellRanger output files  
- store count matrix into a Seurat Object for each culture condition  
- save seurat object for further analyzes

Based on documentations : 
- https://satijalab.org/seurat/articles/multimodal_vignette.html    
- https://github.com/satijalab/seurat/issues/2362
    
As output file, we obtain a list of seurat object (one object per condition). A seurat object is a container gathering data (as raw count matrix) but also analyzes (as PCA or clustering) of a *single-cell dataset*. It as several *Assays* depending the data type : RNA, Protein (ADT), Normalization SCT... Each assay contains its own *count matrix* with features (= genes from GRCh38 database) as rows and cells identified by a unique barcode as column. Principal command lines to handle seurat object are presented in the vignette : https://satijalab.org/seurat/articles/essential_commands.html.  
  
```{r, Working directories and external script}

# Load working directories
directory = str_extract(string = getwd(), pattern = "[:graph:]+(?=bin)")
directory_data = paste0(directory, "data/scRNAseq/")
start_time = Sys.time()

# Create a unique folder for output corresponding to the date of the day
current_date = format(Sys.time(), "%Y%m%d")
dir.create(path = paste0(directory, "exp/scRNAseq_CITEseq_R_CreateSeuratObject/"))
dir.create(path = paste0(directory, "exp/scRNAseq_CITEseq_R_CreateSeuratObject/", current_date))
directory_output = paste0(directory, "exp/scRNAseq_CITEseq_R_CreateSeuratObject/", current_date, "/")

# Load external script with functions and constants
source(file = paste0(directory, "bin/", "functions_constants.R"))

```

```{r, Input loading}

# List of files available in data folder
matrix_list = list.files(path = directory_data, pattern = "mtx")
barcodes_list = list.files(path = directory_data, pattern = "barcodes")
features_list = list.files(path = directory_data, pattern = "features")

# Extract number and names of conditions
if (length(matrix_list) == length(barcodes_list) & 
    length(barcodes_list) == length(features_list)) {
  number_condition = length(matrix_list)
} else { print("Files missing, check data folder") }

list_condition = str_extract(string = matrix_list, pattern = "(?<=scRNAseq_)[:alnum:]+(?=_)")

# Initialize output list 
list_seurat_obj = list()

```

<br> 

We work here from the folder we received from Cochin platform, with all output files from CellRanger software. For each multiplexing sample, we have a folder with all needed files for *count matrix* creation, which are stored in : [sample name]/outs/per_sample_outs/sample[1ou2]/count. For each simple sample obtained with updated CellRanger software, files are stored in folders named as : [sample name]/outs/filtered_feature_bc_matrix. To simplify, the files we use in this code were moved and renammed to follow the same nomenclature.  
To easily distinguish the gene (assay = RNA) from the protein (assay = ADT) for the CD34 marker, protein detection is indicated under CD34.1. FOR CD133 marker, there is no problem as the associated gene as a different name known as PROM1.  

```{r, Create Individual Seurat Object - function definition}

create_seurat_ADT_RNA = function(matrix_file, barcode_file, feature_file) {
  
  # Create a global large dgCMatrix with genes, ADT (CITEseq) and HTO (multiplexing) info
  matrix = as.sparse(ReadMtx(
    mtx = paste0(directory_data, matrix_file), # detected RNA
    cells = paste0(directory_data, barcode_file), # detected cells list
    features = paste0(directory_data, feature_file) # detected genes or antibody list
    )) 
  
  # Separate matrix into one gene matrix and one ADT matrix
  tab_matrix = as.data.frame(matrix)
  tab_antibody = tab_matrix %>%
    dplyr::filter(row.names(tab_matrix) %in% c("CD34.1", "CD133"))
  tab_genes = tab_matrix %>%
    dplyr::filter(!row.names(tab_matrix) %in% c("CD34.1", "CD133", "CMO301", "CMO302"))

  # Create seurat object with gene matrix count in assay RNA
  seurat_obj = CreateSeuratObject(counts = tab_genes, assay = "RNA")
  # Create a new assay to store ADT matrix count
  adt_assay = CreateAssayObject(counts = tab_antibody)
  # Add the new assay into the original seurat object
  seurat_obj[["ADT"]] = adt_assay
  # Add key for features of each assay
  Key(seurat_obj[["RNA"]]) = "rna_"
  Key(seurat_obj[["ADT"]]) = "cite_"
  
  return(seurat_obj)
  
}

```

```{r, Create Individual Seurat Object - call function}

# Call function to create objects and store in a list
for (i in 1:number_condition) {

  list_seurat_obj[[list_condition[i]]] = create_seurat_ADT_RNA(
    matrix_file = str_subset(string = matrix_list, 
                             pattern = paste0(list_condition[i], "_")),
    barcode_file =  str_subset(string = barcodes_list, 
                               pattern = paste0(list_condition[i], "_")),
    feature_file = str_subset(string = features_list, 
                              pattern = paste0(list_condition[i], "_"))
  )

  # Update orig.ident with the condition
  Idents(object = list_seurat_obj[[i]]) = names(list_seurat_obj[i])
  list_seurat_obj[[i]]$orig.ident = names(list_seurat_obj[i])
  
  print(paste(list_condition[i], "seurat obj created"))

}

```

```{r, Create Individual Seurat Object - check outputs}

# Check each objects : number of assays and CITEseq antibody detected
for (i in 1:number_condition) {
  
  cat(names(list_seurat_obj)[i], " : \n",
      "- Assays :", Assays(list_seurat_obj[[i]]), "\n",
      "- CITEseq :", rownames(list_seurat_obj[[i]][["ADT"]]), "\n",
      "- Nb cells :", ncol(list_seurat_obj[[i]]), "\n",
      "- Nb features RNA :", nrow(list_seurat_obj[[i]][["RNA"]]), "\n",
      "- Nb features ADT :", nrow(list_seurat_obj[[i]][["ADT"]]), "\n")
}

# Visualize first object of the list to see the structure
## global structure
list_seurat_obj["CTRL"]
## meta.data 
data.frame(list_seurat_obj[["CTRL"]]@meta.data[1:10,]) %>% kable()
## RNA assay with the genes
list_seurat_obj[["CTRL"]]@assays$RNA
## raw count matrix store in counts.slot (first 10 rows and column)
data.frame(list_seurat_obj[["CTRL"]]@assays$RNA@counts[1:10, 1:5]) %>% kable()
## normalized count matrix store in data.slot (first 10 rows and column)
# => same as the raw matrix because the normalization is not done yet
data.frame(list_seurat_obj[["CTRL"]]@assays$RNA@data[1:10, 1:5]) %>% kable()

```
  
When creating seurat objects, the raw counts are stored in both the counts and data slot at the beginning. 
Then, the NormalizeData() function will work off of the counts slots and overwrite the data slot when the counts slot will still contain the raw data.

```{r, Save outputs}

saveRDS(list_seurat_obj, file = paste0(directory_output, "list_seurat_obj.rds"))

```

```{r, Rsession}

end_time = Sys.time()
cat("Total execution time : ", as.numeric (end_time - start_time, units = "mins"), "minutes")

# Clean working space and memory 
rm(list = ls())
gc()

# Show package version
sessionInfo()

```

