---
title: "scRNAseq_CITEseq_R_FilterData"
author: "Laëtitia Racine"
date: "06/12/2021"
subtitle: "Last modification : `r format(Sys.time(), '%d %B, %Y')`"
output:
  html_document:
    code_folding: "hide"
    toc: true
    toc_float: false
    theme: journal
---

<style>
body {text-align: justify}
</style>

```{r, Setup, include=FALSE}

knitr::opts_chunk$set(echo = TRUE)
knitr::opts_chunk$set(warning = FALSE)
knitr::opts_chunk$set(message = FALSE)

```

```{r, Dependencies}

library(Seurat)
library(dplyr)
library(ggplot2)
library(kableExtra)
library(gridExtra) # function grid.arrange()
library(grid) # function textGrob()
library(matrixStats) # function rowMedians()
library(knitr)
library(stringr)

```



# Overview 

This code aims at :        
- filtering out cells based on their number of detected genes    
- filtering out cells based on their mitochondrial RNA percentage    
- filtering out cells based on their ribosomal RNA percentage    
- filtering out genes to remove mitochondrial ones    
- summarizing the number of cells/genes rejected by each filter    
- analyzing the data structure      
  
Based on documentations :         
- https://satijalab.org/seurat/articles/pbmc3k_tutorial.html       
- https://nbisweden.github.io/workshop-scRNAseq/labs/compiled/seurat/seurat_01_qc.html         
        
We obtain as **output documents** : a list of filtered seurat objects (one per condition) with all the cumulative filters applied, a list of lists (one list per condition) with seurat objects after each filter and multiple summary dataframes. Careful, the filters are cumulative which means that the code has to be launched manually to apply only one filter at a time without the others.  
  
```{r, Working directories and external script}

# Load working directories
directory = str_extract(string = getwd(), pattern = "[:graph:]+(?=bin)")
start_time = Sys.time()

# Create a unique folder for output corresponding to the date of the day
current_date = format(Sys.time(), "%Y%m%d")
dir.create(path = paste0(directory, "exp/scRNAseq_CITEseq_R_FilterData/"))
dir.create(path = paste0(directory, "exp/scRNAseq_CITEseq_R_FilterData/", current_date))
directory_output = paste0(directory, "exp/scRNAseq_CITEseq_R_FilterData/", current_date, "/")

# Load external script with functions and constants
source(file = paste0(directory, "bin/", "functions_constants.R"))

```

```{r, Input loading and output initializing}

# Load seurat objects (initial object without any filter)
dir = pic_last_dir(paste0(directory, "exp/", "scRNAseq_CITEseq_R_CreateSeuratObject/"))
list_seurat_obj = readRDS(paste0(dir, "/", "list_seurat_obj.rds"))
number_conditions = length(list_seurat_obj)

# Initialize empty list for output (list_seurat_obj with cumulative filters)
filtered_list_seurat_obj = list()

# Initialize a list of list to store all versions of seurat object (different filter levels)
list_seurat_obj_filter = list()
for(i in 1:length(list_seurat_obj)) { list_seurat_obj_filter[[i]] = list()}
names(list_seurat_obj_filter) = names(list_seurat_obj)

```



<br><br><br>



# Apply CUMULATIVE quality control filters on the dataset 


## Computation of mitochondrial and ribosomal RNA percentage

```{r, Mitochondrial and ribosomal percentage}

ARN_percentage_fun = function(obj) {
  
  # Extract mitochondrial and ribosomal gene list
  mito_genes = rownames(obj)[grep("^MT-", rownames(obj))]
  ribo_genes <- rownames(obj)[grep("^RP[SL]", rownames(obj))]
  cat("\n", "Mitochondrial genes", length(mito_genes), ":", mito_genes, "\n")
  cat("Ribosomal genes", length(ribo_genes), ":", ribo_genes, "\n")

  # Add percentages in seurat object metadata
  obj[["percent.mt"]] <- PercentageFeatureSet(obj, pattern = "^MT-")
  obj[["percent.ribo"]] <- PercentageFeatureSet(obj, pattern = "^RP[SL]")

  # Output 
  return(obj)
  
}

for (i in 1:number_conditions) {
  # Calculate percentage of mitochondrial and ribosomal RNA = call function
  filtered_list_seurat_obj[[i]] = ARN_percentage_fun(list_seurat_obj[[i]])
  # Save seurat object without filter in list_seurat_obj_filter
  list_seurat_obj_filter[[i]]["seurat_obj"] = filtered_list_seurat_obj[[i]]
}
  
names(filtered_list_seurat_obj) = names(list_seurat_obj)

```


## Cells - Filter_nbG : keep cells with **1000 to 6500 detected genes**

Cells in which few genes were detected could be the result of a technical bias. We decided to apply a filter targeting a minimal number of genes to allow them to be considered as “true cells”. With the same logic, the opposite problem could arise when too many genes are detected giving the cells the appearance of doublet. It exists a percentage of error relative to the cells encapsulated compared to the number inserted in the chip. It is therefore possible for two cells to be encapsulated in the same GEM bead; in this case their transcripts will have the same barcode and will be considered as belonging to the same one cell. All this considered, it is advised to clean the data set by applying filters to remove any doublets. We can use tools before to merge the different data set to do it (https://www.youtube.com/watch?v=NqvAS4HgmrE) but we can also just apply filters relative to the number of genes detected per cell (https://kb.10xgenomics.com/hc/en-us/articles/360001074271-Does-Cell-Ranger-automatically-exclude-doublets-). It’s what we do here. We can therefore apply those two filters in order to keep the most interesting cells from a qualitative point of view. To know where to apply the filters, we consider the number of genes detected per cell according to the different conditions.

```{r, Number of genes detected per cell, results='hold', fig.width=20, fig.height=12}

# Initialize list of graphs 
plot_feature_graphs = list()

# Loop to create features_RNA graphs and store them into the list
for (i in 1:number_conditions) {

  plot_feature = VlnPlot(filtered_list_seurat_obj[[i]], features = "nFeature_RNA") + 
    geom_hline(aes(yintercept = 6500), color = "red", size = 1) +
    geom_hline(aes(yintercept = 1000), color = "red", size = 1) +
    stat_summary(fun = median, geom='point', size = 5 , colour = "white") +
    ylim(0,8500) +
    scale_fill_manual(values = color_code, limits = force) +
    theme(axis.title.x = element_blank(),
          plot.title = element_blank()) +
    NoLegend() 
 
  plot_feature_graphs[[i]] = plot_feature
  
}

# Show graphs
grid.arrange(grobs = plot_feature_graphs,
             ncol = 5,
             nrow = 2,
             top = textGrob("nFeature_RNA"))

```

<br>
Based on those graphics, we can place a threshold. By doing so we decide to disregard all the cells with less than *1000 genes and those with more than 6500 genes*. The median number for genes detected per cell is between 2731 and 3880 genes depending on the condition cf. results websummary CellRanger), this is a positive indication to the threshold we established. We can be quite drastic with this threshold since we are not looking for a rare population. 

```{r, Filter_nbG}

for (i in 1:number_conditions) {
  # Apply filter on cell depending on their genes number 
  filtered_list_seurat_obj[[i]] = subset(filtered_list_seurat_obj[[i]], 
                                         subset = 1000 < nFeature_RNA & 
                                                  nFeature_RNA < 6500)
  filtered_list_seurat_obj[[i]][["filter_nbG"]] = "1000<nFeatures>6500"
  # Save seurat object with nbG filter in list_seurat_obj_filter
  list_seurat_obj_filter[[i]]["seurat_obj_nbG"] = filtered_list_seurat_obj[[i]] 
}

```


## Cells - Filter_pMT & Filter_pRB : keep cells with **less than 25% of mRNA** and **more than 5% of rRNA**

Cells presenting a high percentage of mitochondrial RNA and a lower percentage of ribosomal RNA are either dead or dying. When we trace the scatter plot percent.ribo vs percent.mt, we can see two distinct populations: a first one on the left with a low percentage of of mitochondrial RNA and a percentage of ribosomal RNA between 10 and 50% indicating living cells, and a second population on the right characterized by a high percentage of mitochondrial RNA and a lower percentage of ribosomal RNA. In order to remove most of the dead cells and still maintain a sufficient amount of data we apply a filter with the following parameters: keep the cells **less than 25% of mitochondrial RNA** and **more than 5% of ribosomal RNA**. 

```{r, Percentages RNA, fig.width=16, fig.height=20, fig.fullwidth=TRUE}

plot_percentage_graphs = list()

for (i in 1:number_conditions) {
  
  p1 = VlnPlot(filtered_list_seurat_obj[[i]], 
               features = "percent.mt", 
               group.by = "orig.ident") +
    geom_hline(aes(yintercept = 25), color = "red", size = 1) +
    ylim(0,100)  +
    scale_fill_manual(values = color_code, limits = force) +
    theme(axis.title.x = element_blank()) +
    NoLegend()

  p2 = VlnPlot(filtered_list_seurat_obj[[i]], 
               features = "percent.ribo", 
               group.by = "orig.ident") +
    geom_hline(aes(yintercept = 5), color = "red", size = 1) +
    ylim(0,100) +
    scale_fill_manual(values = color_code, limits = force) +
    theme(axis.title.x = element_blank()) +
    NoLegend()

  p3 = FeatureScatter(filtered_list_seurat_obj[[i]], 
                      feature1 = "percent.mt", 
                      feature2 = "percent.ribo", 
                      group.by = "orig.ident") +
    ylim(0,100) +
    scale_color_manual(values = color_code, limits = force) +
    annotate("rect", xmin = 0, xmax = 25, ymin = 5, 
             ymax = 60, alpha = 0.1, color = "red") +
    ggtitle(names(filtered_list_seurat_obj[i])) +
    NoLegend()

  plot_percentage_graphs[[i]] = list(p1, p2, p3)
  
}

# Show graphs
grid.arrange(grobs = unlist(plot_percentage_graphs, recursive = FALSE),
             ncol = 6,
             top = textGrob("Mitochondrial and ribosomal RNA percentages"))

```

```{r, Filter_pMT_pRB}

for (i in 1:number_conditions) {
  
  # Apply filter on cell depending on their percentage of mitochondrial RNA
  filtered_list_seurat_obj[[i]] = subset(filtered_list_seurat_obj[[i]], subset = percent.mt < 25)
  filtered_list_seurat_obj[[i]][["filter_pMT"]] = "< 25% MT" 
  
  # Save seurat object with nbG and pMT filters
  list_seurat_obj_filter[[i]]["seurat_obj_nbG_pMT"] = filtered_list_seurat_obj[[i]] 
  
  # Apply filter on cell depending on their percentage of ribosomal RNA 
  filtered_list_seurat_obj[[i]] = subset(filtered_list_seurat_obj[[i]], subset = percent.ribo > 5)
  filtered_list_seurat_obj[[i]][["filter_pRB"]] = "> 5% RB"  

  # Save seurat object with nbG, pMT and pRB filters
  list_seurat_obj_filter[[i]]["seurat_obj_nbG_pMT_pRB"] = filtered_list_seurat_obj[[i]] 
  
}

```


## Genes - Filter_rmMT : **mitochondrial** genes are removed

To avoid cells clustering on the basis of mitochondrial genes we decide to disregard them from our data set. 

```{r, Filter_rmMT}

rm_mito_list = function(obj) {
  # Extract features "ADT" names
  DefaultAssay(obj) = "ADT"
  feature_adt_list = rownames(obj)
  # Extract features "RNA gene" names and identify mitochondrial ones
  DefaultAssay(obj) = "RNA"
  mito_genes = rownames(obj)[grep("^MT-", rownames(obj))]
  feature_rna_list = rownames(obj)
  # Keep ADT and non mitochondrial RNA genes
  feature_list = c(feature_adt_list, feature_rna_list)
  # The elements of setdiff(x,y) are those in x but not in y.
  features_keep = setdiff(feature_list, mito_genes) 
  return(features_keep)
}


for (i in 1:number_conditions) {
  # Apply filter on genes to remove mitochondrial ones
  filtered_list_seurat_obj[[i]] = subset(filtered_list_seurat_obj[[i]], 
                                         features = rm_mito_list(filtered_list_seurat_obj[[i]]))
  filtered_list_seurat_obj[[i]][["filter_rmMT"]] = "rm MTgenes" 
  # Save seurat object with nbG, pMT, pRB and rmMT filters
  list_seurat_obj_filter[[i]]["seurat_obj_nbG_pMT_pRB_rmMT"] = filtered_list_seurat_obj[[i]] 
}

```

## Genes - Filter_nbCG : keep genes detected in more than 3 cells

As we are not searching for rare population, we decided to remove genes expressed in less than 3 cells. 

```{r, Filter_rmCG}

rm_undetected_genes= function(obj) {
  # Extract features "ADT" names
  DefaultAssay(obj) = "ADT"
  feature_adt_list = rownames(obj)
  # Extract features "RNA gene" names and identify genes detected in less than 3 cells
  DefaultAssay(obj) = "RNA"
  features_genes_list = rownames(obj)
  features_genes_tab = as.data.frame(obj@assays$RNA@counts) %>% dplyr::mutate_all(as.logical) 
  features_genes_tab = features_genes_tab %>% dplyr::mutate(rowSums = rowSums(features_genes_tab))
  undetected_genes = features_genes_tab %>% dplyr::filter(rowSums < 3)
  undetected_genes_list = rownames(undetected_genes)
  # The elements of setdiff(x,y) are those in x but not in y.
  features_keep = setdiff(features_genes_list, undetected_genes_list) 
  # Keep ADT too
  features_keep = c(features_keep, feature_adt_list)
  return(features_keep)
}

for (i in 1:number_conditions) {
  # Apply filter on genes to remove mitochondrial ones
  filtered_list_seurat_obj[[i]] = subset(filtered_list_seurat_obj[[i]], 
                                         features = rm_undetected_genes(filtered_list_seurat_obj[[i]]))
  filtered_list_seurat_obj[[i]][["filter_rmMT"]] = "rm genes detected in <= 3 cells" 
  # Save seurat object with nbG, pMT, pRB and rmMT filters
  list_seurat_obj_filter[[i]]["seurat_obj_nbG_pMT_pRB_rmMT_rmCG"] = filtered_list_seurat_obj[[i]] 
}

```

```{r, Save outputs part1, fig.show='hide', results='hide'}

ggsave(plot = grid.arrange(grobs = plot_feature_graphs, ncol = 5, 
                           nrow = 2, top = textGrob("nFeature_RNA")),
       filename = paste0(directory_output, "plot_feature_filter.png"),
       width = 16, height = 12)

ggsave(plot = grid.arrange(
  grobs = unlist(plot_percentage_graphs, recursive=F),
  ncol = 6, top = textGrob("Mitochondrial and ribosomal RNA percentages")),
  filename = paste0(directory_output, "plot_percentMTRB_filter.png"),
  width = 24, height = 24)

rm(plot_feature_graphs, plot_percentage_graphs)
rm(rm_mito_list, ARN_percentage_fun, pic_last_dir, loadRData)
rm(p1, p2, p3, dir, directory, plot_feature)
gc()

```



<br><br><br>


# Summarize the number of cells/genes rejected by each filter    

```{r}

# Check each objects : number of assays and CITEseq antibody detected
for (i in 1:number_conditions) {
  
  cat(names(filtered_list_seurat_obj)[i], " : \n",
      "- Assays :", Assays(filtered_list_seurat_obj[[i]]), "\n",
      "- CITEseq :", rownames(filtered_list_seurat_obj[[i]][["ADT"]]), "\n",
      "- Nb cells :", ncol(filtered_list_seurat_obj[[i]]), "\n",
      "- Nb features RNA :", nrow(filtered_list_seurat_obj[[i]][["RNA"]]), "\n",
      "- Nb features ADT :", nrow(filtered_list_seurat_obj[[i]][["ADT"]]), "\n")
}

```


![](/home/rparmentier/Bureau/Git_Differentiation_Metabolism/data/matrix1.png)

```{r, Filter_summary}

# Initialize
filter_tab = data_frame(condition = character(),
                        category = character(),
                        no_filter = integer(),
                        filter_nbG_keep = integer(),
                        filter_nbG_pMT_keep = integer(),
                        filter_nbG_pMT_pRB_keep = integer(),
                        filter_nbG_pMT_pRB_rmMT_keep = integer(),
                        filter_nbG_pMT_pRB_rmMT_rmCG_keep = integer())
 
for(i in 1:number_conditions) {

  temp_gene = data.frame(condition = names(list_seurat_obj[i]),
                         category = "genes",
                         no_filter = nrow(
                           list_seurat_obj_filter[[i]][["seurat_obj"]]@assays$RNA@counts),
                         filter_nbG_keep = nrow(
                           list_seurat_obj_filter[[i]][["seurat_obj_nbG"]]@assays$RNA@counts),
                         filter_nbG_pMT_keep = nrow(
                           list_seurat_obj_filter[[i]][["seurat_obj_nbG_pMT"]]@assays$RNA@counts),
                         filter_nbG_pMT_pRB_keep = nrow(
                           list_seurat_obj_filter[[i]][["seurat_obj_nbG_pMT_pRB"]]@assays$RNA@counts),
                         filter_nbG_pMT_pRB_rmMT_keep = nrow(
                           list_seurat_obj_filter[[i]][["seurat_obj_nbG_pMT_pRB_rmMT"]]@assays$RNA@counts),
                         filter_nbG_pMT_pRB_rmMT_rmCG_keep = nrow(
                           list_seurat_obj_filter[[i]][["seurat_obj_nbG_pMT_pRB_rmMT_rmCG"]]@assays$RNA@counts))
  
  temp_cell = data.frame(condition = names(list_seurat_obj[i]),
                         category = "cells",
                         no_filter = ncol(list_seurat_obj_filter[[i]][["seurat_obj"]]),
                         filter_nbG_keep = ncol(
                           list_seurat_obj_filter[[i]][["seurat_obj_nbG"]]),
                         filter_nbG_pMT_keep = ncol(
                           list_seurat_obj_filter[[i]][["seurat_obj_nbG_pMT"]]),
                         filter_nbG_pMT_pRB_keep = ncol(
                           list_seurat_obj_filter[[i]][["seurat_obj_nbG_pMT_pRB"]]),
                         filter_nbG_pMT_pRB_rmMT_keep = ncol(
                           list_seurat_obj_filter[[i]][["seurat_obj_nbG_pMT_pRB_rmMT"]]),
                         filter_nbG_pMT_pRB_rmMT_rmCG_keep = ncol(
                           list_seurat_obj_filter[[i]][["seurat_obj_nbG_pMT_pRB_rmMT_rmCG"]]))
  
   temp = rbind(temp_gene, temp_cell)
   filter_tab = rbind(filter_tab, temp)
   
}

filter_tab = filter_tab %>%
  dplyr::mutate(filter_nbG_remove = no_filter - filter_nbG_keep) %>%
  dplyr::mutate(filter_pMT_remove = filter_nbG_keep - filter_nbG_pMT_keep) %>%
  dplyr::mutate(filter_pRB_remove = filter_nbG_pMT_keep - filter_nbG_pMT_pRB_keep) %>%
  dplyr::mutate(filter_rmMT_remove = filter_nbG_pMT_pRB_keep - filter_nbG_pMT_pRB_rmMT_keep) %>%
  dplyr::mutate(filter_rmCG_remove = filter_nbG_pMT_pRB_rmMT_keep - filter_nbG_pMT_pRB_rmMT_rmCG_keep)

filter_tab %>% 
  dplyr::filter(category=="genes") %>%
  kable() %>%
  kable_styling(bootstrap_options = c("condensed")) %>%
  row_spec(0:10, color = "black", background = "#FF8A8A") %>%
  scroll_box(width = "900px")

filter_tab %>%
  dplyr::filter(category=="cells") %>%
  kable() %>%
  kable_styling(bootstrap_options = c("condensed")) %>%  
  row_spec(0:10, color = "black", background = "#8AAFFF") %>%
  scroll_box(width = "900px")

```

<br>

Other filters could be applied on the data set. For example, we could apply a filter considering a minimal number of UMI per genes in order to separate them from simple background noise. Another example applies a filter to disregard any gene that monopolizes the reads. Before venturing further, we continue with our analysis which would let us decide which filters would be more pertinent to apply. This avoids the mistake of disregarding too much information from the start only to have to go back and consider them later on. 

```{r, Save outputs part2, results ='hide'}

write.csv2(x = filter_tab, row.names = FALSE,
           file = paste0(directory_output, "filter_tab_seurat_obj.csv"))

saveRDS(object = list_seurat_obj_filter, 
        file = paste0(directory_output, "list_seurat_obj_qc_each_filter.rds"))

rm(filter_tab, list_seurat_obj_filter, extract_nbgene_detected)
rm(temp, temp_cell, temp_gene, list_seurat_obj)
gc()

```


# Data structure visualization

## Genes and cells information

```{r, Cells parameters}

summary_cell_tab = data.frame(condition = character(),
                              statistic = character(),
                              nbgenes_detected_per_cell = integer(),
                              percent_MT_RNA = integer(),
                              percent_RB_RNA = integer())

for(i in 1:number_conditions) {
  # nFeature_RNA : number of genes detected in each cell
  temp = data.frame(condition = rep(names(filtered_list_seurat_obj[i]), 4),
                    statistic = c("min", "median", "mean", "max"),
                    nbgenes_detected_per_cell = c(
                      min(filtered_list_seurat_obj[[i]]$nFeature_RNA), 
                      median(filtered_list_seurat_obj[[i]]$nFeature_RNA), 
                      mean(filtered_list_seurat_obj[[i]]$nFeature_RNA), 
                      max(filtered_list_seurat_obj[[i]]$nFeature_RNA)),
                    percent_MT_RNA = c(
                      min(filtered_list_seurat_obj[[i]]$percent.mt), 
                      median(filtered_list_seurat_obj[[i]]$percent.mt),
                      mean(filtered_list_seurat_obj[[i]]$percent.mt),
                      max(filtered_list_seurat_obj[[i]]$percent.mt)),
                    percent_RB_RNA = c(
                      min(filtered_list_seurat_obj[[i]]$percent.ribo), 
                      median(filtered_list_seurat_obj[[i]]$percent.ribo),
                      mean(filtered_list_seurat_obj[[i]]$percent.ribo),
                      max(filtered_list_seurat_obj[[i]]$percent.ribo)))
  summary_cell_tab = rbind(summary_cell_tab, temp)
  
}

summary_cell_tab %>%
  kable() %>%
  kable_styling()


```

```{r, Extract matrix per condition}

# Save matrix protein per cell - raw data 
list_protein = lapply(filtered_list_seurat_obj, function(x) { 
  as.data.frame(x@assays$ADT@counts) }) 
saveRDS(object = list_protein, file = paste0(directory_output, "list_mat_raw_proteins.rds"))
rm(list_protein)

# extract raw count matrix for each condition
list_mat_raw = lapply(filtered_list_seurat_obj, function(x) { 
  as.data.frame(x@assays$RNA@counts) }) 

# filter matrix to keep only genes detected in at least one cell
list_mat_detected = lapply(list_mat_raw, function(x) {
  detected_genes = rowSums(x) !=0
  detected_genes = detected_genes[detected_genes == TRUE]
  detected_genes = names(detected_genes)
  x = x %>% dplyr::filter(rownames(x) %in% detected_genes)
})

# Save matrix gene UMI per cell - raw data (with undetected genes as well)
saveRDS(object = list_mat_raw, file = paste0(directory_output, "list_mat_raw_genes.rds"))
rm(list_mat_raw)

```

![](/home/rparmentier/Bureau/Git_Differentiation_Metabolism/data/matrix2.png)  

meanUMI_per_gene_corrected : the cells with zero UMI for this gene are discarded for  
this calculation so we have the mean UMI of the cells that expressed the gene studied.  

![](/home/rparmentier/Bureau/Git_Differentiation_Metabolism/data/matrix3.png)

```{r, Genes parameters}

# Extract statistic information
list_ind_genes = list()
for (i in 1:length(list_mat_detected)) {
  
    # Extract matrix 
    temp = list_mat_detected[[i]]
    # Transform into matrix
    temp_mat = as.matrix(temp)
    # Transform into binary matrix (detected/non detected)
    temp_binary = temp %>% mutate_all(as.logical)
    # Transform tab to calcul mean
    temp_mean = temp_mat
    temp_mean[temp_mean == 0] = NA
    # Tab with individual genes info
    ind_genes = data.frame(
      gene = rownames(temp_mat),
      nbcells_expressing_gene = rowSums(temp_binary),
      nbcells_expressing_1UMI = rowSums(temp_mat == 1),
      sumUMI_per_gene = rowSums(temp_mat),
      minUMI_per_gene = rowMins(temp_mat),
      medianUMI_per_gene = rowMedians(temp_mat),
      meanUMI_per_gene = rowMeans(temp_mat),
      meanUMI_per_gene_corrected = round(rowMeans(temp_mean, na.rm = TRUE),3),
      maxUMI_per_gene = rowMaxs(temp_mat))
    ind_genes[is.na(ind_genes)] = 0
       
    list_ind_genes[[i]] = ind_genes
    
}

# Join tabs from different conditions
names(list_ind_genes) = names(list_mat_detected)
ind_genes_tab = dplyr::bind_rows(list_ind_genes, .id = "condition")

# Show tab
ind_genes_tab[1:100,] %>%
  kable(caption = "Extract of the global table") %>%
  kable_styling(bootstrap_options = c("condensed")) %>%
  scroll_box(width = "900px", height = "500px")

```

```{r, Matrix info}

list_summary_cond = list()
list_genes_detected_1cell = list()
list_genes_detected_inf_equal_3cells = list()
list_histo_cells = list()
list_histo_genes = list()

for (i in 1:length(list_mat_detected)) {
  
    # Extract matrix 
    temp = list_mat_detected[[i]]
    # Transform into matrix
    temp_mat = as.matrix(temp)
    # Transform into binary matrix (detected/non detected)
    temp_binary = temp %>% mutate_all(as.logical)

    # Tab summary per condition
    summary_condition = data.frame(
      condition = names(list_mat_detected[i]),
      nbtot_gene = nrow(temp_mat),
      nbtot_cell = ncol(temp_mat),
      # minimum/median/max nb cells expressing simultaneously a gene
      min_nbcells_expressing_same_gene = min(rowSums(temp_binary)), 
      median_nbcells_expressing_same_gene = median(rowSums(temp_binary)),
      max_nbcells_expressing_same_gene = max(rowSums(temp_binary)),
      # minimum/median/max nb genes detected in cells
      min_nbgenes_detected_per_cell = min(colSums(temp_binary)), 
      median_nbgenes_detected_per_cell = median(colSums(temp_binary)),
      max_nbgenes_detected_per_cell = max(colSums(temp_binary)),
      # minimum/median/max sum UMI of all cells for one gene
      min_sumUMI_per_gene = min(rowSums(temp_mat)),
      median_sumUMI_per_gene = median(rowSums(temp_mat)),
      max_sumUMI_per_gene = max(rowSums(temp_mat)),
      # minimum/median/max sum UMI of all genes for one cell
      min_sumUMI_per_cell = min(colSums(temp_mat)),
      median_sumUMI_per_cell = median(colSums(temp_mat)),
      max_sumUMI_per_cell = max(colSums(temp_mat)),
      # details
      nbgenes_expressed_in_allcells = sum((rowSums(temp_binary) == ncol(temp_mat)), na.rm = TRUE),
      nbgenes_1UMI_total = sum((rowSums(temp_mat) == 1), na.rm = TRUE),
      nbgenes_detected_1cell = sum((rowSums(temp_binary) == 1), na.rm = TRUE),
      nbgenes_detected_inf_equal_3cells = sum((rowSums(temp_binary) <= 3), na.rm = TRUE),
      nbcells_1UMI_total = sum((colSums(temp_mat) == 1), na.rm = TRUE),
      nbcells_expressing_only1gene = sum((colSums(temp_binary) == 1), na.rm = TRUE)
    )
    
    list_summary_cond[[i]] = summary_condition
    
    # Extract list of noteworthy genes (maybe to remove later)
    list_genes_detected_1cell[[i]] = rownames(temp_binary[rowSums(temp_binary) == 1,])
    list_genes_detected_inf_equal_3cells[[i]] = rownames(temp_binary[rowSums(temp_binary) <= 3,])
    
    # Visualize distribution
    nbcells_expressing_same_gene = data.frame(nbcells_expressing_same_gene = rowSums(temp_binary))
    nbgenes_detected_per_cell = data.frame(nbgenes_detected_per_cell = colSums(temp_binary))
    
    histo_cells = ggplot(data = nbcells_expressing_same_gene, aes(x=nbcells_expressing_same_gene)) +
      geom_histogram(binwidth = 20, fill = "#FF7A7A", col = "#A11313") +
      ggtitle(names(list_mat_detected[i]))
    histo_genes = ggplot(data = nbgenes_detected_per_cell, aes(x=nbgenes_detected_per_cell)) +
      geom_histogram(binwidth = 20, fill = "#7AB5FF", col = "#1351A1") +
      ggtitle(names(list_mat_detected[i]))

    list_histo_cells[[i]] = histo_cells
    list_histo_genes[[i]] = histo_genes
    
}

names(list_summary_cond) = names(list_mat_detected)
summary_cond_tab = dplyr::bind_rows(list_summary_cond, .id = "condition")

```

![](/home/rparmentier/Bureau/Git_Differentiation_Metabolism/data/matrix4.png)

```{r, fig.width=18, fig.height=18}

summary_cond_tab[ ,1:9] %>%
  kable() %>%
  kable_styling(bootstrap_options = c("condensed")) %>% 
  column_spec(4:6, color = "black", background = "#FF8A8A") %>%
  column_spec(7:9, color = "black", background = "#8AAFFF") %>%
  scroll_box(width = "900px")

do.call("grid.arrange", c(list_histo_cells, ncol=3))
do.call("grid.arrange", c(list_histo_genes, ncol=3))

```

![](/home/rparmentier/Bureau/Git_Differentiation_Metabolism/data/matrix5.png)

```{r}

summary_cond_tab[ ,c(1:3, 10:15)] %>%
  kable() %>%
  kable_styling(bootstrap_options = c("condensed")) %>% 
  column_spec(4:6, color = "black", background = "#FF8A8A") %>%
  column_spec(7:9, color = "black", background = "#8AAFFF") %>%
  scroll_box(width = "900px") 

summary_cond_tab[ ,c(1:3, 16:ncol(summary_cond_tab))] %>%
  kable() %>%
  kable_styling(bootstrap_options = c("condensed")) %>% 
  scroll_box(width = "900px")

```

```{r}

write.csv2(x = summary_cell_tab, row.names = FALSE,
           file = paste0(directory_output, "summary_cell_tab.csv"))
write.csv2(x = summary_cond_tab, row.names = FALSE,
           file = paste0(directory_output, "summary_cond_tab.csv"))
write.csv2(x = ind_genes_tab, row.names = FALSE,
           file = paste0(directory_output, "individual_gene_tab.csv"))

# Save Seurat object with all filters
saveRDS(object = filtered_list_seurat_obj, 
        file = paste0(directory_output, "list_seurat_obj_qc.rds"))

# Save list of genes (possible to filter out them)
saveRDS(list_genes_detected_1cell, file = paste0(directory_output, "list_genes_detected_1cell.rds"))
saveRDS(list_genes_detected_inf_equal_3cells, file = paste0(directory_output, "list_genes_detected_inf_equal_3cells.rds"))


rm(summary_cell_tab, summary_cond_tab, ind_genes_tab)
rm(filtered_list_seurat_obj, list_genes_detected_1cell)
rm(list_genes_detected_inf_equal_3cells, list_histo_cells)
rm(histo_cells, histo_genes, ind_genes, list_histo_genes)
rm(nbcells_expressing_same_gene, nbgenes_detected_per_cell)
rm(summary_condition, temp_binary, temp, temp_mat, temp_mean)
rm(list_ind_genes, list_summary_cond)

```

## UMI distribution

```{r}

tab_UMI = data.frame(condition = character(),
                     category = character(),
                     sumUMI_equal_1 = integer(),
                     sumUMI_equal_2 = integer(),
                     sumUMI_between_2_10 = integer(),
                     sumUMI_between_10_100 = integer(),
                     sumUMI_between_100_1000 = integer(),
                     sumUMI_between_1000_2000 = integer(),
                     sumUMI_between_2000_5000 = integer(),
                     sumUMI_between_5000_10000 = integer(),
                     sumUMI_between_10000_20000 = integer(),
                     sumUMI_sup_20000 = integer())

for (i in 1:length(list_mat_detected)) {
  
  x = list_mat_detected[[i]]
  
  cell_sum = data.frame(sum_UMI = colSums(x)) %>%
    dplyr::mutate(category = case_when(sum_UMI==1 ~ "sumUMI_equal_1",
                                       sum_UMI==2 ~ "sumUMI_equal_2",
                                       (2<=sum_UMI & sum_UMI<10) ~ "sumUMI_between_2_10",
                                       (10<=sum_UMI & sum_UMI<100) ~ "sumUMI_between_10_100",
                                       (100<=sum_UMI  & sum_UMI<1000) ~ "sumUMI_between_100_1000",
                                       (1000<=sum_UMI  & sum_UMI<2000) ~ "sumUMI_between_1000_2000",
                                       (2000<=sum_UMI  & sum_UMI<5000) ~ "sumUMI_between_2000_5000",
                                       (5000<=sum_UMI  & sum_UMI<10000) ~ "sumUMI_between_5000_10000",
                                       (10000<=sum_UMI  & sum_UMI<20000) ~ "sumUMI_between_10000_20000",
                                       sum_UMI>20000 ~ "sumUMI_sup_20000"))
  
  recap_cell = data.frame(table(cell_sum$category))
  recap_cell = tidyr::pivot_wider(recap_cell, names_from = "Var1", values_from = "Freq")
  recap_cell = recap_cell %>% dplyr::mutate(condition = names(list_mat_detected[i]), category = "cells")
  
  tab_UMI = bind_rows(tab_UMI, recap_cell)
  
  gene_sum = data.frame(sum_UMI = rowSums(x)) %>%
    dplyr::mutate(category = case_when(sum_UMI==1 ~ "sumUMI_equal_1",
                                       sum_UMI==2 ~ "sumUMI_equal_2",
                                       (2<=sum_UMI & sum_UMI<10) ~ "sumUMI_between_2_10",
                                       (10<=sum_UMI & sum_UMI<100) ~ "sumUMI_between_10_100",
                                       (100<=sum_UMI  & sum_UMI<1000) ~ "sumUMI_between_100_1000",
                                       (1000<=sum_UMI  & sum_UMI<2000) ~ "sumUMI_between_1000_2000",
                                       (2000<=sum_UMI  & sum_UMI<5000) ~ "sumUMI_between_2000_5000",
                                       (5000<=sum_UMI  & sum_UMI<10000) ~ "sumUMI_between_5000_10000",
                                       (10000<=sum_UMI  & sum_UMI<20000) ~ "sumUMI_between_10000_20000",
                                       sum_UMI>20000 ~ "sumUMI_sup_20000"))
  
  recap_gene = data.frame(table(gene_sum$category))
  recap_gene = tidyr::pivot_wider(recap_gene, names_from = "Var1", values_from = "Freq")
  recap_gene = recap_gene %>% dplyr::mutate(condition = names(list_mat_detected[i]), category = "genes")
  
  tab_UMI = bind_rows(tab_UMI, recap_gene)
  
}

tab_UMI[is.na(tab_UMI)] = 0

tab_UMI %>% dplyr::filter(category == "genes") %>%
  kable() %>%
  kable_styling(bootstrap_options = c("condensed")) %>%
  row_spec(0:10, color = "black", background = "#FF8A8A") %>%
  scroll_box(width = "900px")

tab_UMI %>% dplyr::filter(category == "cells") %>%
  kable() %>%
  kable_styling(bootstrap_options = c("condensed")) %>%  
  row_spec(0:10, color = "black", background = "#8AAFFF") %>%
  scroll_box(width = "900px")

```

```{r, Save outputs, fig.show='hide'}

write.csv2(x = tab_UMI, row.names = FALSE,
           file = paste0(directory_output, "tab_UMI.csv"))

# Save matrix gene UMI per cell - filtered data, only detected genes 
saveRDS(object = list_mat_detected, file = paste0(directory_output, "list_mat_detected_genes.rds"))

```

```{r, Rsession}

end_time = Sys.time()
cat("Total execution time : ", as.numeric (end_time - start_time, units = "mins"), "minutes")

# Show package version
sessionInfo()

```

```{r, results='hide'}

# Clean working space and memory 
rm(list = ls())
gc()

```