---
title: "scRNAseq_CITEseq_R_NormalizeData"
author: "Laëtitia Racine"
date: "15/12/2021"
subtitle: "Last modification : `r format(Sys.time(), '%d %B, %Y')`"
output:
  html_document:
    code_folding: "hide"
    toc: true
    toc_float: false
    theme: journal
---

<style>
body {text-align: justify}
</style>

```{r, Setup, include=FALSE}

knitr::opts_chunk$set(echo = TRUE)
knitr::opts_chunk$set(warning = FALSE)
knitr::opts_chunk$set(message = FALSE)

```

```{r, Dependencies}

library(dplyr)
library(Seurat)
library(ggplot2)
library(gridExtra) # function grid.arrange()
library(grid) # function textGrob()
library(sctransform)
library(matrixStats) # function rowMedians()
library(kableExtra)
library(stringr)

# BiocManager::install("glmGamPoi")

```


# Overview 

This code aims to :        

- I : normalize seurat objects with basic functions from Seurat package (assay : "RNA")  
https://satijalab.org/seurat/articles/pbmc3k_tutorial.html    
https://satijalab.org/seurat/reference/normalizedata      
https://satijalab.org/seurat/reference/findvariablefeatures      
https://satijalab.org/seurat/reference/scaledata      
  
- II : visualize cell cycle influence and apply regression if needed  
https://satijalab.org/seurat/articles/cell_cycle_vignette.html  
https://hbctraining.github.io/scRNA-seq/lessons/06_SC_SCT_and_integration.html  
https://nbisweden.github.io/workshop-scRNAseq/labs/compiled/seurat/seurat_01_qc.html#Calculate_cell-cycle_scores  
https://satijalab.org/seurat/archive/v3.1/cell_cycle_vignette.html    
  
- III : normalize seurat objects with SCTransform (assay : "SCT")  
https://chipster.csc.fi/manual/single-cell-seurat-sctransform.html  
https://satijalab.org/seurat/articles/sctransform_vignette.html  
https://www.biostars.org/p/446554/  
https://cran.r-project.org/web/packages/sctransform/sctransform.pdf  
https://rdrr.io/cran/sctransform/man/vst.html  
  
- IV : normalize data from CITEseq (assay : "ADT)  
https://satijalab.org/seurat/articles/multimodal_vignette.html    
   
It's important to respect the order here. As mentionned in the issue : https://github.com/satijalab/seurat/issues/1679, the standard normalization as to be done first, then the CellCycleScoring and finally the SCT normalization with cell cycle score as vars.to.regress. We obtain as **output documents** a list of seurat objects (one per condition) with or without cell cycle regression and with two normalizations available dependint the assay we work with. For *downstream analysis*, we will mainly work with the "SCT" assay.   
  
```{r, Working directories and external script}

# Load working directories
directory = str_extract(string = getwd(), pattern = "[:graph:]+(?=bin)")
directory_data = paste0(directory, "data/scRNAseq/")
start_time = Sys.time()

# Create a unique folder for output corresponding to the date of the day
current_date = format(Sys.time(), "%Y%m%d")
dir.create(path = paste0(directory, "exp/scRNAseq_CITEseq_R_NormalizeData/"))
dir.create(path = paste0(directory, "exp/scRNAseq_CITEseq_R_NormalizeData/", current_date))
directory_output = paste0(directory, "exp/scRNAseq_CITEseq_R_NormalizeData/", current_date, "/")

# Load external script with functions and constants
source(file = paste0(directory, "bin/", "functions_constants.R"))

```

```{r, Input loading}

# Load seurat objects
dir = pic_last_dir(paste0(directory, "exp/", "scRNAseq_CITEseq_R_FilterData/"))
list_seurat_obj = readRDS(paste0(dir, "/", "list_seurat_obj_qc.rds"))
number_conditions = length(list_seurat_obj)

```


# Normalize "RNA assay" with Seurat package

## NormalizeData()

NormalizeData() with “LogNormalize” method : x is the number of UMI detected for a specific gene in a single cell. To normalize, x is divided by the sum of detected UMI of all the genes in this cell. Then, it's multiplied by the *scaling factor* which default value is 10 000. 1 is added to the result and the neperian logarithm (ln) is calculated (https://github.com/satijalab/seurat/issues/3630).    
"LogNormalize scales the data by library size so it eliminates the confounding effect of different read depth across the cells."  

```{r, NormalizeData}

list_seurat_obj = lapply(list_seurat_obj, function(x) { 
  NormalizeData(object = x, 
                assay = "RNA",
                normalization.method = "LogNormalize",
                scale.factor = 10000,
                verbose = TRUE) 
})

```


## FindVariableFeatures()

FindVariableFeatures() select a subset of interesting genes to keep for downstream analysis. 2000 genes are chosen (default value) among the most variable genes between the cells, i.e. those strongly expressed in some cells but sparsely in others. To extract this genes list, the "vst" method is used : "first, fits a line to the relationship of log(variance) and log(mean) using local polynomial regression. Then standardizes the feature values using the observed mean and expected variance (given by the fitted line). Feature variance is then calculated on the standardized values after clipping to a maximum (see clip.max parameter)".
*Next analysis as UMAP or PCA are based on those 2000 most variable genes.*

```{r, FindVariableFeatures}

list_seurat_obj = lapply(list_seurat_obj, function(x) { 
  FindVariableFeatures(object = x, 
                       assay = "RNA",
                       selection.method = 'vst',
                       nfeatures = "2000",   
                       verbose = TRUE) 
})

```

```{r, MostVariableGenes, fig.width = 20, fig.height = 24, results='hold'}

# Initialize empty output list
features_plot_list = list()
features_gene_list = list()

for (i in 1:number_conditions) {

  name_obj = names(list_seurat_obj)[i]
  variable_genes = VariableFeatures(list_seurat_obj[[i]])
  features_gene_list[[i]] = variable_genes
  
   plot_vfeature = LabelPoints(plot = VariableFeaturePlot(list_seurat_obj[[i]]), 
                               points = head(variable_genes, 20),
                               repel = TRUE,
                               xnudge = 0,
                               ynudge = 0) +
    theme(plot.background = element_rect(fill = "white"),
          legend.position = "none") +
    ggtitle(name_obj)
  features_plot_list[[i]] = plot_vfeature
  
}

grid.arrange(grobs = features_plot_list, ncol = 3)
names(features_plot_list) = names(list_seurat_obj)
names(features_gene_list) = names(list_seurat_obj)

```

## ScaleData()

"Since highly expressed genes exhibit the highest amount of variation and we don’t want our ‘highly variable genes’ only to reflect high expression, we need to scale the data to scale variation with expression level." Therefore, the ScaleData() step gives equal weight in downstream analyses, so that highly-expressed genes do not dominate. It's a linear transformation that shifts the expression of each gene, so that the mean expression across cells is 0 and scales the expression of each gene, so that the variance across cells is 1.   
(https://hbctraining.github.io/scRNA-seq_online/lessons/06_SC_SCT_normalization.html)  
By default, do.center and do.scale are defined as TRUE. "Setting center to TRUE will center the expression for each feature by subtracting the average expression for that feature. Setting scale to TRUE will scale the expression level for each feature by dividing the centered feature expression levels by their standard deviations if center is TRUE and by their root mean square otherwise."     
It is possible to scale the data with a regression (for example, nUMI, or percent.mito....). The first scaling that we apply is without regression but we can do another one latter with regression that will therefore overwrite the output of ScaleData().    
Here we do an individual scaling per condition (split.by = orig.ident). It can potentially highlight genes with great disparity between conditions for the PCA which is what we want to study.   

```{r, ScaleData}

list_seurat_obj = lapply(list_seurat_obj, function(x) { 
  ScaleData(object = x, 
            assay = "RNA",
            split.by = "orig.ident",
            verbose = TRUE) 
})

```


# Cell Cycle influence on cell-to-cell variability 

To study cell cycle, data must be normalized and scaled. Then, a list of specific genes associated to S or G2M phases (stored in the cc.genes Seurat function) is used to set a cell cycle score for each cell to decide in which phase they belong. Markers are anticorrelated, cells without any marker expression are in a phase of cell division or in G1 phase. This step allows to "mitigate the effects of cell cycle heterogeneity in scRNA-seq data by calculating cell cycle phase scores based on canonical markers, and regressing these out of the data during pre-processing".    

```{r}

cat("There are", length(cc.genes$s.genes), 
    "markers associated to S phase :", "\n", 
    cc.genes$s.genes, "\n")

cat("There are", length(cc.genes$g2m.genes), 
    "markers associated to G2M phase :", "\n", 
    cc.genes$g2m.genes, "\n")

```


## Cell cycle score assignment

CellCycleScoring() set the identity of the Seurat object to the cell-cycle phase by passing set.ident = TRUE (the original identities are stored as old.ident).
```{r, results = 'hold', fig.width=20, fig.height=6, fig.fullwidth=TRUE}

list_seurat_obj = lapply(list_seurat_obj, function(x) { 
  CellCycleScoring(object = x, 
                   assay = "RNA",
                   s.features = cc.genes$s.genes,
                   g2m.features = cc.genes$g2m.genes,
                   set.ident = TRUE,
                   verbose = TRUE) 
})

```


## Cells' distribution based on their cell cycle score
```{r}

list_seurat_obj = lapply(list_seurat_obj, function(x) { 
  RunPCA(object = x, 
         assay = "RNA",
         reduction.name = "pca.cellcycle",
         npcs = 50,
         features = c(cc.genes$s.genes, cc.genes$g2m.genes), 
         verbose = TRUE) 
})

tab_cellcycle = data.frame()

for (i in 1:number_conditions) {
  
  name_cond = names(list_seurat_obj[i])
  tab = as.data.frame(table(list_seurat_obj[[i]]$Phase)) %>%
    tidyr::pivot_wider(names_from ="Var1", values_from = "Freq") %>%
    dplyr::mutate(condition = name_cond)
  tab_cellcycle = rbind(tab_cellcycle, tab)
}

tab_cellcycle %>%
  kable() %>%
  kable_styling()

```


## Visualize cell cycle scoring
```{r, fig.width = 14, fig.height = 14, fig.fullwidth=TRUE}

# initialize list 
plot_cc_list = list()

for (i in 1:number_conditions) {
  
  name_obj = names(list_seurat_obj)[i]
  
  plot.s = VlnPlot(object = list_seurat_obj[[i]], features = "S.Score", 
                   group.by = "orig.ident", pt.size = 0.1, 
                   assay = "RNA", y.max = 1.5,
                   same.y.lims = TRUE) + 
    scale_fill_manual(values = color_code, limits = force) +
    ggtitle(paste0(name_obj, " - S.Score")) +
    NoLegend() +
    theme(axis.title.x = element_blank())
  
  plot.g2 = VlnPlot(object = list_seurat_obj[[i]], features = "G2M.Score", 
                    group.by = "orig.ident", pt.size = 0.1,
                    assay = "RNA", y.max = 1.5,
                    same.y.lims = TRUE) + 
    scale_fill_manual(values = color_code, limits = force) +
    ggtitle(paste0(name_obj, " - G2M.Score")) +
    NoLegend() +
    theme(axis.title.x = element_blank())
  
  cc_cond_pca = DimPlot(object = list_seurat_obj[[i]], 
                        reduction = "pca.cellcycle",
                        group.by = "Phase",
                        split.by = "orig.ident") + ggtitle(name_obj)
    
  plot_cc_list[[i]] = list(plot.s, plot.g2, cc_cond_pca)
  
}


grid.arrange(grobs = unlist(plot_cc_list, recursive = FALSE),
             ncol = 6,
             top = textGrob("Cell cycle scoring repartition"))

```


## Number of UMI involved in cell cycle genes
```{r}

# Matrix (gene/cell/UMInumber) per condition
list_mat = lapply(list_seurat_obj, function(x) { 
  as.data.frame(x@assays$RNA@counts) 
}) 

# Keep only cell cycle genes
print("Filter genes to keep only cell cycle genes")
list_mat = lapply(list_mat, function(x) { 
  x %>% dplyr::filter(row.names(x) %in% c(cc.genes$s.genes, cc.genes$g2m.genes)) 
})
  
# Extract the number of cell expressing each gene and
# statistics on UM (sum, mean, median, variance)
print("Calculate tab statistics")
list_mat = lapply(list_mat, function(x) {

    temp_nbcells = x
    ncol_temp_nbcells = ncol(temp_nbcells)
    temp_nbcells[] = lapply(temp_nbcells, as.logical)
    temp_nbcells = temp_nbcells %>% 
      dplyr::mutate(nbcells = rowSums(.)) %>%
      dplyr::mutate(percentcells = round((nbcells/ncol_temp_nbcells) *100)) %>%
      tibble::rownames_to_column(var = "gene") %>%
      dplyr::select(gene, nbcells, percentcells)

    temp_umi = x 
    temp_umi = temp_umi %>%
      dplyr::mutate(totalUMI = rowSums(.),
                    medianUMI = rowMedians(as.matrix(temp_umi[,-totalUMI])),
                    meanUMI = round(rowMeans(as.matrix(temp_umi[,c(-totalUMI, -medianUMI)])), 2),
                    varUMI = round(rowVars(as.matrix(temp_umi[,c(-totalUMI, -medianUMI, -meanUMI)])), 2)) %>%
      tibble::rownames_to_column(var = "gene") %>%
      dplyr::mutate(phase = ifelse(gene %in% cc.genes$s.genes, "S", "G2M")) %>% 
      dplyr::select(gene, phase, totalUMI, meanUMI, medianUMI, varUMI)
    x = full_join(temp_nbcells, temp_umi, by = "gene")
    return(x)
  })

# Bind conditions in a global table
tab_cc = dplyr::bind_rows(list_mat, .id = "condition")

tab_cc %>%
      kable() %>%
      kable_styling() %>%
      scroll_box(height = "400px", width = "100%")

```

```{r, results = 'hold', fig.width=22, fig.height=14, fig.fullwidth=TRUE}

plot_smedian = ggplot(tab_cc %>% filter(phase == "S"), 
                      aes(x=gene, y=medianUMI, fill=condition)) +
  geom_col(position = "dodge") +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5)) +
  scale_fill_manual(values = color_code, limits = force) +
  ggtitle("Phase S genes - Median UMI")

plot_smean = ggplot(tab_cc %>% filter(phase == "S"), 
                    aes(x=gene, y=meanUMI, fill=condition)) +
    geom_col(position = "dodge") +
    theme(axis.text.x = element_text(angle = 90, vjust = 0.5)) +
    scale_fill_manual(values = color_code, limits = force) +
    ggtitle("Phase S genes - Mean UMI")
  
plot_g2median = ggplot(tab_cc %>% filter(phase == "G2M"), 
                       aes(x=gene, y=medianUMI, fill=condition)) +
    geom_col(position = "dodge") +
    theme(axis.text.x = element_text(angle = 90, vjust = 0.5)) +
    scale_fill_manual(values = color_code, limits = force) +
    ggtitle("Phase G2M genes - Median UMI")

plot_g2mean = ggplot(tab_cc %>% filter(phase == "G2M"), 
                     aes(x=gene, y=meanUMI, fill=condition)) +
    geom_col(position = "dodge") +
    theme(axis.text.x = element_text(angle = 90, vjust = 0.5)) +
    scale_fill_manual(values = color_code, limits = force) +
    ggtitle("Phase G2M genes - Mean UMI")

print(plot_smedian/plot_smean | plot_g2median/plot_g2mean)

```

<br>

We can decide to regress the data based on the cell cycle score : this regression is possible during the SCT normalization (see above). 

# Normalize "RNA assay" with SCTransform (new assay "SCT")

SCTransform function gathers filters on data, a normalization, the extraction of the 3000 most variable genes and the scaling. To improve computing speed, the "glmGamPoi" method is used in the SCTransform function. New data are stored in a different assay "SCT". "Compared to basic Seurat normalisation + scaling + variable gene finding, the SCTransform workflow performs more effective normalization, strongly removing technical effects from the data. This allows more PCs to be selected in the PCA step, as the higher PCs are more likely to represent subtle, but biologically relevant, sources of heterogeneity instead of variation in sequencing depth."    
<br>
*Filters*: Careful ! Default filters parameters are "hiden" in the function.   
- Filter out cells which have less than this many genes expressed [200] => our own filter is more restrictive (1000 genes) so no impact  
- **Filter out cells which have higher unique gene count [2500] => much more restritif than ours but I don't know how to remove it**  
- Filter out cells which have higher mitochondrial transcript percentage [5]     
By putting min_cells = 1, we remove the min_cells	filter that only uses genes that have been detected in at least this many cells; default is 5.  
As we already have a bunch of filters applied during preprocessing step, we try to remove as possible the default filters of the function. 
We consulted the parameters of vst function (included in SCTransform) and added them in the parameters of the SCTransform function to control the filters as much as possible : https://rdrr.io/cran/sctransform/man/vst.html   
<br>
*Normalization* : Expression values are normalized using the SCTransform normalization method, which uses Pearson residuals from “regularized negative binomial regression,” where cellular sequencing depth is utilized as a covariate in a generalized linear model (GLM). The parameters for the model are estimated by pooling information across genes that are expressing at similar levels. This should remove the technical characteristics but preserve the biological heterogeneity, and avoid overfitting the model to the data. 

```{r, Normalize RNA SCT}

# Be careful, we can't do both normalization in the same object 
# because the second one erases the first one

# SCT normalization with cell cycle regression
list_seurat_obj_cc = lapply(list_seurat_obj, function(x) { 
  SCTransform(object = x, 
              assay = "RNA", 
              new.assay.name = "SCT",
              method = "glmGamPoi", 
              do.center = TRUE, 
              do.scale = FALSE,
              variable.features.n = 3000,
              vars.to.regress = c("S.Score", "G2M.Score"),
              min_cells = 1,
              ncells = NULL,
              verbose = FALSE) 
})

# SCT normalization without cell cycle regression
list_seurat_obj = lapply(list_seurat_obj, function(x) { 
  SCTransform(object = x, 
              assay = "RNA", 
              new.assay.name = "SCT",
              method = "glmGamPoi", 
              do.center = TRUE, 
              do.scale = FALSE,
              variable.features.n = 3000,
              vars.to.regress = NULL,
              min_cells = 1,
              ncells = NULL,
              verbose = FALSE)
  
})

```


# Normalize "ADT assay" with Seurat package

Normalization method CLR applies a centered log ratio transformation, with parameter margin = 2 the normalisation is performing across cells.

```{r, Normalize ADT}

list_seurat_obj = lapply(list_seurat_obj, function(x) {
  NormalizeData(object = x, 
                normalization.method = "CLR",
                margin = 2,
                assay = "ADT")
})

list_seurat_obj_cc = lapply(list_seurat_obj_cc, function(x) {
  NormalizeData(object = x, 
                normalization.method = "CLR",
                margin = 2,
                assay = "ADT")
})

```

```{r, Save outputs, fig.show='hide'}

ggsave(plot = grid.arrange(grobs = features_plot_list, ncol = 3),
       filename = paste0(directory_output, "plot_2000variablesfeatures.svg"),
       width = 28, height = 24)

ggsave(plot = grid.arrange(grobs = unlist(plot_cc_list, recursive = FALSE),
                           ncol = 6, top = textGrob("Cell cycle scoring repartition")),
       filename = paste0(directory_output, "plot_cellcyclescoring.svg"),
       width = 28, height = 24)

ggsave(plot = (plot_smedian/plot_smean | plot_g2median/plot_g2mean), 
       filename = paste0(directory_output, "plot_cellcycle_statUMI.svg"),
       height = 6, width = 20)
  
write.csv2(x = tab_cellcycle, file = paste0(directory_output, "tab_cellcycle.csv"))
write.csv2(x = tab_cc, file = paste0(directory_output, "tab_genes_cellcycle.csv"))

saveRDS(list_seurat_obj_cc, file = paste0(directory_output, "list_seurat_obj_qc_norm_ccr.rds"))
saveRDS(list_seurat_obj, file = paste0(directory_output, "list_seurat_obj_qc_norm.rds"))

```

```{r, Rsession}

end_time = Sys.time()
cat("Total execution time : ", as.numeric (end_time - start_time, units = "mins"), "minutes")

# Clean working space and memory 
rm(list = ls())
gc()

# Show package version
sessionInfo()

```
