---
title: "scRNAseq_CITEseq_R_DWNS_PathwaysAnalysis"
author: "Laëtitia Racine"
date: "2023-01-25"
subtitle: "Last modification : `r format(Sys.time(), '%d %B, %Y')`"
output:
  html_document:
    code_folding: "hide"
    toc: true
    toc_float: false
    theme: journal
---

<style>
body {text-align: justify}
</style>

```{r setup, include=FALSE}

knitr::opts_chunk$set(echo = TRUE)
knitr::opts_chunk$set(warning = FALSE)
knitr::opts_chunk$set(message = FALSE)

```

```{r, Global dependencies}

library(Seurat)
library(dplyr)
library(stringr)
library(ggplot2)
library(kableExtra)
library(gridExtra)
library(ggrepel) # geom_text_repel() function
library(RColorBrewer)
library(cowplot) # plot_grid() function
library(viridis)

```

```{r, Working directories and external script}

# Load working directories
directory = str_extract(string = getwd(), pattern = "[:graph:]+(?=bin)")
start_time = Sys.time()

# Create a unique folder for output corresponding to the date of the day
current_date = format(Sys.time(), "%Y%m%d")
dir.create(path = paste0(directory, "exp/scRNAseq_CITEseq_R_DWNS_PathwaysAnalysis/"))
dir.create(path = paste0(directory, "exp/scRNAseq_CITEseq_R_DWNS_PathwaysAnalysis/", current_date))
directory_output = paste0(directory, "exp/scRNAseq_CITEseq_R_DWNS_PathwaysAnalysis/", current_date, "/")

# Load external script with functions and constants
source(file = paste0(directory, "bin/", "functions_constants.R"))

```

# Overview 

```{r, Input loading}

dir = pic_last_dir(paste0(directory, "exp/", "scRNAseq_CITEseq_R_ReductionDimension/"))
merged_prep = readRDS(file = paste0(dir, "/", "merged_all_prep_1-40dim.rds"))

combination = list(c("CTRL","AOA"), c("CTRL","2DG"), c("CTRL","DON"),
                   c("CTRL", "VPA"), c("CTRL","CTRLaK"), c("2DG","2DGaK"),
                   c("DON","DONaK"), c("AOA","AOAaK"), c("CTRLaK","2DGaK"),
                   c("CTRLaK","AOAaK"), c("CTRLaK","DONaK"))

```

As a reminder, here are the clusters we identified in the dataset.

```{r, Show clusters, fig.width = 6, fig.height = 4}

DimPlot(merged_prep,
        label = TRUE,
        reduction = "umap",
        group.by = "clusters")

```



<br><br><br>



# Gene set generation : identify interesting pathways to analyze 

To perform pathway analysis on our scRNAseq dataset, we need to compare our genes expression to gene set (list of genes connected to pathway). To do so, there are multiple packages avaliable depending on the final result we want : ReactomeGSA, SCPA, GAGE, Pathview.... We decided to try two of them, ReactomeGSA and SCPA. To do so, the first step is to extract the gene set to allow the comparaison and the gene set enrichment analysis. Based, on documentation, we tried two packages (ReactomeContentService4R and msigdbr) to extract gene set from Reactome database. We compared the output and select the best for further analysis.  
- https://bioconductor.org/packages/devel/bioc/vignettes/ReactomeContentService4R/inst/doc/ReactomeContentService4R.html  
- https://jackbibby1.github.io/SCPA/articles/quick_start.html  
- https://github.com/igordot/msigdbr/issues/4  
- https://www.gsea-msigdb.org/gsea/msigdb/human/genesets.jsp?collection=CP:REACTOME  
  
## Extract list of pathways with package ReactomeContentService4R

```{r}
library(ReactomeContentService4R)
```

```{r}

## Fetch all human pathway available in the Reactome database

pathways_db = getSchemaClass(class= "Pathway", species= "human", all= TRUE)
pathways_db = data.frame(pathways_db) %>% dplyr::select(displayName, stId)
colnames(pathways_db) = c("pathway_name", "pathway_id")

pathways_db %>%
  kable() %>%
  kable_styling() %>%
  scroll_box(height = "250px", width = "100%")

## Extract genes associated to the pathways of the Reactome database

genes_pathways = list()
for (i in 1:nrow(pathways_db)) {
  #cat("Extract pathway", i, "/", nrow(pathways_db), "\n")
  path = pathways_db$pathway_id[i]
  genes = event2Ids(path)$geneSymbol
  if (is.null(genes)) {
    cat("No gene associated to pathway", path, "\n")
  } else {
      genes_pathways[[path]] = data.frame(pathway_id = path, gene = genes)
    }
}
genes_pathways = do.call(rbind, genes_pathways)

genes_pathways %>%
  kable() %>%
  kable_styling() %>%
  scroll_box(height = "250px", width = "100%")

```

```{r}

# write.csv2(x = pathways_db, row.names = FALSE,
#            file = paste0(directory_output, "pathways_database_reactome.csv"))
# write.csv2(x = genes_pathways, row.names = FALSE,
#            file = paste0(directory_output, "pathways_genes_reactome.csv"))

pathways_db = read.csv2(paste0(directory_output, "pathways_database_reactome.csv"))
genes_pathways = read.csv2(paste0(directory_output, "pathways_genes_reactome.csv"))

```

## Extract list of pathways with package msigdbr

```{r}
library(msigdbr)
```

```{r}

pathways_db_ms <- msigdbr(species = "Homo sapiens", 
                          category = "C2",
                          subcategory = "CP:REACTOME") 

pathways_db_ms %>%
  kable() %>%
  kable_styling() %>%
  scroll_box(height = "250px", width = "100%")

```

## Compare packages 

```{r}

cat("The ReactomeContentService4R allowed us to extract : ", 
    nrow(pathways_db), "gene sets from the Reactome database. \n",
    "The msigdbr allowed us to extract :", 
    length(unique(pathways_db_ms$gs_name)), "gene sets from the Reactome database. \n") 

# intersection
common = intersect(pathways_db_ms$gs_exact_source, pathways_db$pathway_id)
# elements in pathways_db_ms not in pathways_db
diff_msigdbr = setdiff(pathways_db_ms$gs_exact_source, pathways_db$pathway_id)
# elements in pathways_db not in pathways_db_ms
diff_react = setdiff(pathways_db$pathway_id, pathways_db_ms$gs_exact_source)

cat("We found", length(common), "pathways in common between the two lists. \n",
    "Only", length(diff_msigdbr), "pathway from msigdbr is not in the bigger list of pathways from ReactomeContent.",
    "ReactomeContent allow us to extract", length(diff_react), "more pathways. \n",
    "We will now compare the common gene set to see if they are equals.")

# compare gene list for one gene set
path = pathways_db_ms$gs_exact_source[[1]]
if(path %in% pathways_db$pathway_id) { print("TRUE")}
genes_msigdbr = (pathways_db_ms %>% dplyr::filter(gs_exact_source == path))$gene_symbol
genes_reactome = (genes_pathways %>% dplyr::filter(pathway_id == path))$gene 
intersect(genes_msigdbr, genes_reactome)
setdiff(genes_msigdbr, genes_reactome)
setdiff(genes_reactome, genes_msigdbr)
cat("For pathway", path, "we have", length(genes_msigdbr),
    "genes associated for msigdbr package and", length(genes_reactome),
    "genes associated for reactome package.", 
    "All the genes from msigdbr list are in the reactome list.", 
    "The others names are associated proteins or synonyms of the gene.")
# G22P1 = XRCC6
# G22P2 = XRCC5
# HMGIY est la protéine codée par HMGA1
# BAF = BCRG1 = BANF1
# DFS70 = LEDGF = PSIP2 = PSIP1 

```

To summarize, by accessing the same database (Reactome) but with different packages (msigdbr and ReactomeContentService4R), the results are not totally equals. Msigdbr package is less precise and we are missing some information (pathways and gene synonyms) so we decided to use the ReactomeContentService4R gene set list for further analysis. We just have to format the data in a specific way (one list par pathway containing a tab with pathway name/id in column 1 and gene name in column 2).

```{r}

pathways_db_list = list()

for (i in 1:length(unique(genes_pathways$pathway_id))){
  #cat(i, "/", length(unique(genes_pathways$pathway_id)), "\n")
  path = unique(genes_pathways$pathway_id)[i]
  temp = genes_pathways %>% dplyr::filter(pathway_id == path)
  colnames(temp) = c("Pathway", "Genes")
  pathways_db_list[[path]] = temp
}

```

```{r}

saveRDS(object = pathways_db_list, file = paste0(directory_output, "pathways_database_reactome_format_scpa.rds"))

```

## Select interesting pathways for our study 

```{r}

pathways_of_interest = c("R-HSA-15869","R-HSA-70171","R-HSA-70221","R-HSA-70263",
                         "R-HSA-70268","R-HSA-70326","R-HSA-70350","R-HSA-70370",
                         "R-HSA-70635","R-HSA-70688","R-HSA-70895","R-HSA-70921",
                         "R-HSA-71032","R-HSA-71064","R-HSA-71240","R-HSA-71262",
                         "R-HSA-71288","R-HSA-71291","R-HSA-71336","R-HSA-71384",
                         "R-HSA-71387","R-HSA-71403","R-HSA-71406","R-HSA-73817",
                         "R-HSA-75105","R-HSA-77286","R-HSA-77288","R-HSA-77289",
                         "R-HSA-163685","R-HSA-174403","R-HSA-189445","R-HSA-189451",
                         "R-HSA-189483","R-HSA-352230","R-HSA-2262752","R-HSA-9711097",
                         "R-HSA-9711123","R-HSA-3247509","R-HSA-5334118","R-HSA-156581",
                         "R-HSA-156582","R-HSA-212165","R-HSA-8964539")

pathways_interest_tab = pathways_db %>% 
  dplyr::filter(pathway_id %in% pathways_of_interest) %>%
  dplyr::mutate(Pathway = paste(pathway_name, pathway_id, sep =" - "))

pathways_interest_tab %>%
  kable() %>%
  kable_styling() %>%
  scroll_box(height = "250px", width = "100%")

```

```{r}

write.csv2(x = pathways_interest_tab, row.names = FALSE,
           file = paste0(directory_output, "pathways_interest_reactome.csv"))

```

```{r}

# Extract all gene names associated to each pathway
# Keep only gene name present in our dataset  (lots of synonyms to remove)
genes_interest_tab = genes_pathways %>%
  dplyr::filter(pathway_id %in% pathways_of_interest) %>%
  dplyr::filter(gene %in% rownames(merged_prep))
 
genes_interest_tab %>%
  kable() %>%
  kable_styling() %>%
  scroll_box(height = "250px", width = "100%")

```

```{r}

write.csv2(x = genes_interest_tab, row.names = FALSE,
           file = paste0(directory_output, "pathways_genes_interest_reactome.csv"))

```

```{r}
rm(directory, loadRData, GO_clusterProfiler_fun, dir, pathways_db_ms,
   common, diff_msigdbr, diff_react, path, genes_msigdbr, genes_reactome,
   temp)
```


<br><br><br>



# Pathways analysis with ReactomeGSA

```{r}
library(ReactomeGSA)
```

Documentation :    
- https://bioconductor.org/packages/release/bioc/vignettes/ReactomeGSA/inst/doc/analysing-scRNAseq.html  
- https://www.sciencedirect.com/science/article/pii/S1535947620600159  
- https://bioconductor.org/packages/devel/bioc/vignettes/ReactomeGSA/inst/doc/using-reactomegsa.html  
- https://rdrr.io/bioc/ReactomeGSA/man/analyse_sc_clusters.html  
- https://rdrr.io/bioc/ReactomeGSA/src/R/perform_analysis.R  
- https://gsea-msigdb.github.io/ssGSEA-gpmodule/v10/index.html    
  
Workflow  :  Analyses cell groups (idents) of a single-cell RNA-sequencing experiment to get pathway-level expressions for every group of cells.    
- Idents() function => defines which grouping variable to use for the analysis (conditions/clusters...)      
- analyse_sc_clusters() function => returns an ReactomeAnalysisResult object          
- pathways() function => returns a dataframe with pathway-level expression values per cell groups      
- max_difference homemade function => finds the most relevant pathways    
- plots function => plots barplots and heatmaps    

<br>

## Pathway-level expression values

**Call analyse_sc_clusters function on groups of cells based on clustering algorithm**
```{r}

DefaultAssay(merged_prep) = "SCT"
Idents(merged_prep) = "clusters"

gsva_clust = analyse_sc_clusters(
  object = merged_prep,
  use_interactors = TRUE,
  include_disease_pathways = FALSE,
  create_reactome_visualization = FALSE,
  verbose = TRUE
)

gsva_clust

```

```{r}

saveRDS(gsva_clust, paste0(directory_output, "RGSA_gsva_result_clusters.rds"))

```

**Call analyse_sc_clusters function on groups of cells based on culture conditions**
```{r}

DefaultAssay(merged_prep) = "SCT"
Idents(merged_prep) = "orig.ident"

gsva_cond = analyse_sc_clusters(
  object = merged_prep,
  use_interactors = TRUE,
  include_disease_pathways = FALSE,
  create_reactome_visualization = TRUE,
  verbose = TRUE
)

gsva_cond

```

```{r}

saveRDS(gsva_cond, paste0(directory_output, "RGSA_gsva_result_condition.rds"))

```

**Extract pathway-level expression tabs for clusters and conditions**
```{r}

pathway_exp_clust = pathways(gsva_clust)
colnames(pathway_exp_clust) = gsub("\\.Seurat", "", colnames(pathway_exp_clust))
colnames(pathway_exp_clust) = str_replace(string = colnames(pathway_exp_clust),
                                           pattern = "X",
                                           replacement = "cluster_")
pathway_exp_clust = pathway_exp_clust %>%
  tibble::rownames_to_column("pathway_id") %>%
  dplyr::select(pathway_id, Name, cluster_0, cluster_1, cluster_2, cluster_3,
                cluster_4, cluster_5, cluster_6, cluster_7, cluster_8,
                cluster_9, cluster_10, cluster_11, cluster_12, cluster_13,
                cluster_14, cluster_15, cluster_16)

pathway_exp_clust %>%
  kable(caption = "Pathway-level expression for clusters") %>%
  kable_styling() %>%
  scroll_box(height = "250px", width = "100%")

pathway_exp_cond = pathways(gsva_cond)
colnames(pathway_exp_cond) = gsub("\\.Seurat", "", colnames(pathway_exp_cond))
colnames(pathway_exp_cond) = str_replace(string = colnames(pathway_exp_cond),
                                           pattern = "X",
                                           replacement = "")
pathway_exp_cond = pathway_exp_cond %>%
  tibble::rownames_to_column("pathway_id") 

pathway_exp_cond %>%
  kable(caption = "Pathway-level expression for conditions") %>%
  kable_styling() %>%
  scroll_box(height = "250px", width = "100%")

```

```{r}

write.csv2(pathway_exp_clust, row.names = FALSE,
           paste0(directory_output, "RGSA_pathway_clusters.csv"))
write.csv2(pathway_exp_cond, row.names = FALSE,
           paste0(directory_output, "RGSA_pathway_conditions.csv"))

```

**Extract fold-change tabs for clusters and conditions**
```{r}

fc_clust = get_result(x = gsva_clust, type = "fold_changes", name ="Seurat")
colnames(fc_clust) = str_replace(string = colnames(fc_clust),
                                           pattern = "X",
                                           replacement = "cluster_")
fc_clust %>%
  kable(caption = "Fold change for clusters") %>%
  kable_styling() %>%
  scroll_box(height = "250px", width = "100%")

fc_cond = get_result(x = gsva_cond, type = "fold_changes", name ="Seurat")
colnames(fc_cond) = str_replace(string = colnames(fc_cond),
                                           pattern = "X",
                                           replacement = "")
fc_cond %>%
  kable(caption = "Fold change for conditions") %>%
  kable_styling() %>%
  scroll_box(height = "250px", width = "100%")

```

```{r}

write.csv2(fc_cond, row.names = FALSE,
           paste0(directory_output, "RGSA_foldchange_conditions.csv"))
write.csv2(fc_clust, row.names = FALSE,
           paste0(directory_output, "RGSA_foldchange_clusters.csv"))

```

**Visualize outliers groups with PCA**
```{r, fig.width = 16}

gsva_result = gsva_clust
p_clust = plot_gsva_pca(object = gsva_result, pathway_ids = NULL)
p_clust

gsva_result = gsva_cond
p_cond = plot_gsva_pca(object = gsva_cond, pathway_ids = NULL)
p_cond

```

```{r}

ggsave(plot = p_clust, filename = paste0(directory_output, "RGSA_outliers_clusters.svg"))
ggsave(plot = p_cond, filename = paste0(directory_output, "RGSA_outliers_conditions.svg"))

```

The analyse_sc_cluster() function includes several steps :  
- calculate the mean gene expression for every group of cells (idents)  
- map gene identifiers to the database to extract associate pathway(s) (one pathway = gene set) 
- submit this data to a gene set variation analysis (ssGSEA = single sample gene set enrichment analysis => calculates separate enrichment scores for each pairing of a sample (= group of cells) and gene set (=pathway). Each ssGSEA enrichment score represents the degree to which the genes in a particular gene set are coordinately up or down regulated within a sample.

<br>

## Study the most relevant pathways 

**Calcul max difference tab for clusters and conditions comparisons**  
We start with a table containing one pathway-expression value per group of cells (idents = cluster or condition). For each pathway, we identify the maximal and minimal value among the groups and we calculate the difference. Then, we keep the pathway with the highest values of difference because it concerns pathway with high heterogeneity of expression between grups of cells.

```{r}

max_diff_fun = function(pathway_exp_tab, gsva) {
  
  max_difference = do.call(rbind, apply(pathway_exp_tab, 1, function(row) {
    values = as.numeric(row[3:length(row)])
    return(data.frame(pathway_ID = row[1], name = row[2], min = min(values), max = max(values)))
  }))
  
  max_difference$diff = max_difference$max - max_difference$min
  max_difference = max_difference[order(max_difference$diff, decreasing = T), ]
  return(max_difference)
  
}

```

```{r}

relevant_path_clust = max_diff_fun(pathway_exp_clust, gsva_clust)
relevant_path_cond = max_diff_fun(pathway_exp_cond, gsva_cond)

relevant_path_clust %>%
  kable() %>%
  kable_styling() %>%
  scroll_box(height = "250px", width = "100%")

relevant_path_cond %>%
  kable() %>%
  kable_styling() %>%
  scroll_box(height = "250px", width = "100%")

```

**Visualize the 30 most relevant pathways for clusters and conditions**
```{r}

barplots_fun = function(max_difference, gsva, nb_pathway) {
  
  list_plot = list()
  
  for (i in 1:nb_pathway) {
    
    pathway_id = max_difference$pathway_ID[i]
    plot_title = paste0(max_difference$name[i], " (", pathway_id, ")")
    
    plot = plot_gsva_pathway(gsva, pathway_id = pathway_id) +
      labs(title = str_wrap(plot_title, 50)) +
      theme(title = element_text(size = 9))
    
    list_plot[[i]] = plot
    
  }
  
  return(list_plot)

}

```

```{r, fig.width = 14, fig.height = 16}

relevant_path_clust_plots = barplots_fun(relevant_path_clust, gsva_clust, 30)
do.call("grid.arrange", c(relevant_path_clust_plots, ncol=4))

relevant_path_cond_plots = barplots_fun(relevant_path_cond, gsva_cond, 30)
do.call("grid.arrange", c(relevant_path_cond_plots, ncol=4))

```

```{r, fig.width = 10, fig.height = 8}

plot_gsva_heatmap(gsva_clust,
                  pathway_ids = relevant_path_clust_plots$pathway_ID[1:30],
                  margins = c(6,35),
                  dendrogram = "col",
                  scale = "row",
                  lmat = rbind(c(0,3),c(2,1),c(0,4)),
                  lwid = c(0.1,4),
                  lhei = c(1.5,4,1),
                  truncate_names = FALSE,
                  key = TRUE,
                  keysize = 0.5,
                  cexRow=1)

plot_gsva_heatmap(gsva_cond,
                  pathway_ids = relevant_path_cond_plots$pathway_ID[1:30],
                  margins = c(6,35),
                  dendrogram = "col",
                  scale = "row",
                  lmat = rbind(c(0,3),c(2,1),c(0,4)),
                  lwid = c(0.1,4),
                  lhei = c(1.5,4,1),
                  truncate_names = FALSE,
                  key = TRUE,
                  keysize = 0.5,
                  cexRow=1)  https://bioconductor.org/packages/release/bioc/vignettes/ReactomeGSA/inst/doc/analysing-scRNAseq.html

```

```{r, fig.show='hide', results='hide'}

ggsave(plot = do.call("grid.arrange", c(relevant_path_clust_plots, ncol=6)), 
       filename = paste0(directory_output, "RGSA_relevant_path_clust.svg"),
       height = 20, width = 24)
ggsave(plot = do.call("grid.arrange", c(relevant_path_cond_plots, ncol=6)), 
       filename = paste0(directory_output, "RGSA_relevant_path_cond.svg"),
       height = 20, width = 24)


pdf(file = paste0(directory_output, "RGSA_heatmap_clust.pdf"),
    height = 12, width = 16)
plot_gsva_heatmap(gsva_clust,
                  pathway_ids = relevant_path_clust_plots$pathway_ID[1:30],
                  margins = c(6,35),
                  dendrogram = "col",
                  scale = "row",
                  lmat = rbind(c(0,3),c(2,1),c(0,4)),
                  lwid = c(0.1,4),
                  lhei = c(1.5,4,1),
                  truncate_names = FALSE,
                  key = TRUE,
                  keysize = 0.5,
                  cexRow=1)
dev.off()

pdf(file = paste0(directory_output, "RGSA_heatmap_cond.pdf"),
    height = 12, width = 16)
plot_gsva_heatmap(gsva_cond,
                  pathway_ids = relevant_path_cond_plots$pathway_ID[1:30],
                  margins = c(6,35),
                  dendrogram = "col",
                  scale = "row",
                  lmat = rbind(c(0,3),c(2,1),c(0,4)),
                  lwid = c(0.1,4),
                  lhei = c(1.5,4,1),
                  truncate_names = FALSE,
                  key = TRUE,
                  keysize = 0.5,
                  cexRow=1) 
dev.off()

```

<br>

## Study our list of interesting pathways 

We can only study pathways with enough genes represented in our dataset so we have to filter the list of pathway of interest to keep only those we can study.

```{r, fig.width = 14, fig.height = 18}

pathway_int_clust = pathway_exp_clust %>% dplyr::filter(pathway_id %in% pathways_of_interest)
list_int_path_clust = list() 

for (i in 1:nrow(pathway_int_clust)) {
    pathway_id = pathway_int_clust$pathway_id[i]
    plot_title = paste0(pathway_int_clust$Name[i], " (", pathway_id, ")")
    plot = plot_gsva_pathway(gsva_clust, pathway_id = pathway_id) +
      labs(title = str_wrap(plot_title, 50)) +
      theme(title = element_text(size = 9))
    list_int_path_clust[[pathway_id]] = plot
}

do.call("grid.arrange", c(list_int_path_clust, ncol = 4))

pathway_int_cond = pathway_exp_cond %>% dplyr::filter(pathway_id %in% pathways_of_interest)
list_int_path_cond = list() 

for (i in 1:nrow(pathway_int_cond)) {
    pathway_id = pathway_int_cond$pathway_id[i]
    plot_title = paste0(pathway_int_cond$Name[i], " (", pathway_id, ")")
    plot = plot_gsva_pathway(gsva_cond, pathway_id = pathway_id) +
      labs(title = str_wrap(plot_title, 50)) +
      theme(title = element_text(size = 9))
    list_int_path_cond[[pathway_id]] = plot
  }

do.call("grid.arrange", c(list_int_path_cond, ncol = 4))

```

```{r, fig.width = 10, fig.height = 8}

plot_gsva_heatmap(gsva_clust,
                  pathway_ids = pathway_int_clust$pathway_id,
                  margins = c(6,35),
                  dendrogram = "col",
                  scale = "row",
                  lmat = rbind(c(0,3),c(2,1),c(0,4)),
                  lwid = c(0.1,4),
                  lhei = c(1.5,4,1),
                  truncate_names = FALSE,
                  key = TRUE,
                  keysize = 0.5,
                  cexRow=1) 


plot_gsva_heatmap(gsva_cond,
                  pathway_ids = pathway_int_cond$pathway_id,
                  margins = c(6,35),
                  dendrogram = "col",
                  scale = "row",
                  lmat = rbind(c(0,3),c(2,1),c(0,4)),
                  lwid = c(0.1,4),
                  lhei = c(1.5,4,1),
                  truncate_names = FALSE,
                  key = TRUE,
                  keysize = 0.5,
                  cexRow=1) 

```

```{r, fig.show='hide'}

ggsave(plot = do.call("grid.arrange", c(list_int_path_clust, ncol = 4)), 
       filename = paste0(directory_output, "RGSA_interest_path_clust.svg"))
ggsave(plot = do.call("grid.arrange", c(list_int_path_cond, ncol = 4)), 
       filename = paste0(directory_output, "RGSA_interest_path_cond.svg"))


pdf(paste0(directory_output, "RGSA_heatmap_interest_clust.pdf"),
    height = 12, width = 16)
plot_gsva_heatmap(gsva_clust,
                  pathway_ids = pathway_int_clust$pathway_id,
                  margins = c(6,35),
                  dendrogram = "col",
                  scale = "row",
                  lmat = rbind(c(0,3),c(2,1),c(0,4)),
                  lwid = c(0.1,4),
                  lhei = c(1.5,4,1),
                  truncate_names = FALSE,
                  key = TRUE,
                  keysize = 0.5,
                  cexRow=1) 
dev.off()

pdf(paste0(directory_output, "RGSA_heatmap_interest_cond.pdf"),
    height = 12, width = 16)
plot_gsva_heatmap(gsva_cond,
                  pathway_ids = pathway_int_cond$pathway_id,
                  margins = c(6,35),
                  dendrogram = "col",
                  scale = "row",
                  lmat = rbind(c(0,3),c(2,1),c(0,4)),
                  lwid = c(0.1,4),
                  lhei = c(1.5,4,1),
                  truncate_names = FALSE,
                  key = TRUE,
                  keysize = 0.5,
                  cexRow=1) 
dev.off()

```



<br><br><br>



# Pathway analysis with SCPA

"It’s a different approach to pathway analysis that defines pathway activity as a change in multivariate distribution of a given pathway across conditions, rather than enrichment or over representation of genes."    
"This analysis provides a statistic (Qval), which indicates the size of distribution change for a given pathway and can be used for the ranking of pathways in order of biological relevance. This method therefore provides a fundamentally different definition of pathway activity when compared with current tools that typically rely on enrichment of overrepresentation of genes in a given pathway."  
  
Documentation :    
- https://jackbibby1.github.io/SCPA/    
- https://jackbibby1.github.io/SCPA/articles/quick_start.html    
- https://jackbibby1.github.io/SCPA/articles/seurat_comparison.html   
- https://jackbibby1.github.io/SCPA/reference/compare_seurat.html    
- https://doi.org/10.1186/s13059-019-1790-4    
- https://jackbibby1.github.io/SCPA/reference/plot_rank.html    
- https://www.sciencedirect.com/science/article/pii/S2211124722015716    
- https://github.com/jackbibby1/SCPA/blob/HEAD/R/PlotRank.R  

```{r}
library(SCPA)
```

We use the gene set list stored in pathways_db_list.

<br>

## Calculate scpa_out for each combination of conditions

```{r}

list_scpa_cond = list()

for (i in 1:length(combination)) {

  print(combination[[i]])

  scpa_out = compare_seurat(seurat_object = merged_prep,
                           assay = "SCT",
                           group1 = "orig.ident",
                           group1_population = combination[[i]],
                           pathways = pathways_db_list,
                           downsample = 500,
                           min_genes = 5,
                           max_genes = 500,
                           parallel = TRUE,
                           cores = 8) %>%
    dplyr::mutate(color = case_when(FC > 5 & adjPval < 0.01 ~ '#6dbf88', # positive enrichment + good pvalue
                                    FC < -5 & adjPval < 0.01 ~ 'mediumseagreen', # negative enrichment + good pvalue
                                    FC < 5 & FC > -5 & adjPval < 0.01 ~ '#84b0f0', # no enrichment + good pavalue
                                    FC < 5 & FC > -5 & adjPval > 0.01 ~ 'black')) %>% # no enrichment + bad pvalue
     dplyr::mutate(combination = paste(combination[[i]], collapse = "_"))

  scpa_out = left_join(scpa_out, (pathways_db %>% dplyr::rename("Pathway" = "pathway_id")), by = "Pathway") %>%
    dplyr::mutate(pathway_id = Pathway) %>%
    dplyr::mutate(Pathway = paste(pathway_name, Pathway, sep =" - ")) %>%
    dplyr::select(-pathway_name)

  list_scpa_cond[[paste(combination[[i]], collapse = "_")]] = scpa_out

}

```

```{r}

saveRDS(object = list_scpa_cond, file = paste0(directory_output, "SCPA_list_scpa_conditions.rds"))

```

Output :   
- qval : higher qval translates to larger pathway differences between conditions even without enrichment      
- FC enrichment score : is generated from a running sum of mean changes in gene expression from all genes of the pathway  
- FC > 0  => average pathway expression population 1 > population 2  
- FC < 0 => average pathway expression pathway population 1 < population 2  

<br>

## Plot the results 

**Heatmap on interesting pathways for each combination of conditions**
```{r, fig.height = 14, fig.width = 8, results = 'hold'}

tab_heat_scpa_cond = Reduce(full_join, list_scpa_cond) %>%
  dplyr::select(Pathway, qval, combination) %>% 
  tidyr::pivot_wider(names_from = combination, values_from = qval)
colnames(tab_heat_scpa_cond) = c("Pathway", paste(colnames(tab_heat_scpa_cond[,2:ncol(tab_heat_scpa_cond)]),"qval",sep="_"))

plot_ht = plot_heatmap(scpa_out = tab_heat_scpa_cond, 
                       highlight_pathways = pathways_interest_tab$Pathway,
                       column_names = colnames(tab_heat_scpa_cond[,2:ncol(tab_heat_scpa_cond)]),
                       row_fontsize = 10,
                       column_fontsize = 10,
                       show_row_names = FALSE,
                       hm_colors = NULL)
plot_ht

tab_heat_scpa_cond = tab_heat_scpa_cond %>% dplyr::filter(Pathway %in% pathways_interest_tab$Pathway)

plot_ht_int = plot_heatmap(scpa_out = tab_heat_scpa_cond, 
                           highlight_pathways = NULL,
                           column_names = colnames(tab_heat_scpa_cond[,2:ncol(tab_heat_scpa_cond)]),
                           row_fontsize = 8,
                           column_fontsize = 10,
                           show_row_names = TRUE,
                           hm_colors = NULL)
plot_ht_int

```

```{r}

pdf(paste0(directory_output, "SCPA_heatmap_interest_path.pdf"),
    width = 14, height = 10)
plot_heatmap(scpa_out = tab_heat_scpa_cond, 
             highlight_pathways = NULL,
             column_names = colnames(tab_heat_scpa_cond[,2:ncol(tab_heat_scpa_cond)]),
             row_fontsize = 8,
             column_fontsize = 10,
             show_row_names = TRUE,
             hm_colors = NULL)
dev.off()

```


**Enrichment and rank plot for each combination of conditions**
```{r, fig.width = 16, fig.height = 20, results='hold'}

list_enrich_cond_all = list()
list_enrich_cond_int = list()
list_rank_cond = list()

for (i in 1:length(list_scpa_cond)) {
  
  # Enrichment plot with all pathways
  scpa_out = list_scpa_cond[[i]]
  
  plot = ggplot(scpa_out, aes(x = FC, y = qval)) +
    geom_vline(xintercept = c(-5, 5), linetype = "dashed", col = 'black', lwd = 0.3) +
    geom_hline(yintercept = 5, linetype = "dashed", col = 'black', lwd = 0.3) +
    geom_point(alpha = 0.6, cex = 2.6, shape = 21, fill = scpa_out$color, stroke = 0.3) +
    geom_text_repel(data = scpa_out %>% dplyr::filter(qval > 5),
                    aes(label = pathway_id), size = 2, direction = "both",
                    force = 3, force_pull = 2, max.overlaps = 200) +
    coord_cartesian(xlim=c(-20,80), ylim=c(0,11)) +
    xlab("Enrichment") +
    ylab("Qval") +
    ggtitle(names(list_scpa_cond[i])) +
    theme(panel.background = element_blank(),
          panel.border = element_rect(fill = NA),
          aspect.ratio = 1) 
 
  list_enrich_cond_all[[i]] = plot
  
  # Rank plot with interesting pathways 
  scpa_out = list_scpa_cond[[i]] %>% dplyr::mutate(Pathway = pathway_id)
  
  plot_rank = plot_rank(scpa_out = scpa_out,
                        pathway = pathways_interest_tab$pathway_id,
                        population_name = "qval",
                        base_point_size = 4,
                        base_point_color = "gray70",
                        highlight_point_size = 3,
                        highlight_point_color = "cornflowerblue",
                        label_pathway = TRUE,
                        label_size = 2.5)

  list_rank_cond[[i]] = plot_rank
  
  # Enrichment plot with interesting pathways only 
  scpa_out = list_scpa_cond[[i]] %>% dplyr::filter(pathway_id %in% pathways_of_interest) 
  
  plot_int = ggplot(scpa_out, aes(FC, qval)) +
    geom_vline(xintercept = c(-5, 5), linetype = "dashed", col = 'black', lwd = 0.3) +
    geom_hline(yintercept = 5, linetype = "dashed", col = 'black', lwd = 0.3) +
    geom_point(cex = 2.6, shape = 21, fill = scpa_out$color, stroke = 0.3) +
    geom_text_repel(data = scpa_out %>% dplyr::filter(qval > 5),
                    aes(label = pathway_id), size = 2, direction = "both",
                    force = 3, force_pull = 2, max.overlaps = 200) +
    coord_cartesian(xlim=c(-20,80), ylim=c(0,11)) +
    xlab("Enrichment") +
    ylab("Qval") +
    ggtitle(names(list_scpa_cond[i])) +
    theme(panel.background = element_blank(),
          panel.border = element_rect(fill = NA),
          aspect.ratio = 1) 
  
  list_enrich_cond_int[[i]] = plot_int
  
}

do.call("grid.arrange", c(list_enrich_cond_all, ncol = 4))
do.call("grid.arrange", c(list_enrich_cond_int, ncol=4))
do.call("grid.arrange", c(list_rank_cond, ncol = 4))

```

```{r, fig.show="hide"}

ggsave(plot = do.call("grid.arrange", c(list_enrich_cond_all, ncol = 4)),
       filename = paste0(directory_output, "SCPA_enrich_cond.svg"),
       width = 20, height = 14)
ggsave(plot = do.call("grid.arrange", c(list_enrich_cond_int, ncol=4)),
       filename = paste0(directory_output, "SCPA_interest_enrich_cond.svg"),
       width = 20, height = 14)
ggsave(plot = do.call("grid.arrange", c(list_rank_cond, ncol = 4)),
       filename = paste0(directory_output, "SCPA_rank_cond.svg"),
       width = 20, height = 14)

```


<br><br><br>



# "Manual" pathways and genes analysis

Documentation :  
- https://satijalab.org/seurat/reference/addmodulescore  
- https://www.waltermuskovic.com/2021/04/15/seurat-s-addmodulescore-function/  
- https://github.com/satijalab/seurat/issues/1342  
- https://github.com/satijalab/seurat/issues/261  
- https://github.com/satijalab/seurat/issues/4881  
- https://github.com/satijalab/seurat/issues/2495  

<br>

## AddModuleScore with gene set from Reactome database

```{r, results='hold'}

list_modulescore_clust = list()
list_modulescore_cond = list()

for (i in 1:length(unique(genes_interest_tab$pathway_id))) {
  
  pathway = unique(genes_interest_tab$pathway_id)[i]
  tab_pathway = genes_interest_tab %>% dplyr::filter(pathway_id == pathway)
  ms.features = list(as.vector(tab_pathway$gene))
  name = str_extract(string = pathway, pattern = "(?<=R-HSA-)[:digit:]+")
  name = paste0("RHSA", name, "_")
  
  merged_prep = AddModuleScore(object = merged_prep, 
                               features = ms.features, 
                               pool = NULL, #default
                               nbin = 24, #default
                               ctrl = 5, #vignette example
                               k = FALSE, #default
                               assay = "RNA",
                               name = name,
                               seed = 42,
                               search = TRUE)
  
  plot_clust = FeaturePlot(object = merged_prep,
                           features = paste0(name, "1"),
                           reduction = "umap",
                           cols = rev(brewer.pal(n = 11, name = "RdBu")))
  
  plot_cond = FeaturePlot(object = merged_prep,
                          features = paste0(name, "1"),
                          reduction = "umap",
                          split.by = "orig.ident",
                          keep.scale = "feature",
                          combine = FALSE, 
                          cols = rev(brewer.pal(n = 11, name = "RdBu"))) 
  
  list_modulescore_clust[[pathway]] = plot_clust
  list_modulescore_cond[[pathway]] = plot_cond 

}

```

```{r, fig.width=14, fig.height=26, results='hold'}

do.call("grid.arrange", c(list_modulescore_clust, ncol = 4))

```

```{r, fig.width=14, fig.height=10, results='hold'}

for (i in 1:length(list_modulescore_cond)){
  print(plot_grid(plotlist = list_modulescore_cond[[i]], ncol = 4))
}

```

```{r, fig.show='hide'}

ggsave(plot = do.call("grid.arrange", c(list_modulescore_clust, ncol = 4)),
       filename = paste0(directory_output, "modulescore_clust.svg"))

saveRDS(object = list_modulescore_cond, 
        file = paste0(directory_output, "list_modulescore_cond.rds"))
saveRDS(object = list_modulescore_clust, 
        file = paste0(directory_output, "list_modulescore_clust.rds"))

```

<br>

## Visualize expression of specific genes manually selected

```{r}

plot_gene_fun = function(seurat_obj, list, choice){
  
  for (i in 1:length(list)) {
    
    gene = list[i]
    
    if (choice == "clust") {
      
      plot_cl = FeaturePlot(object = seurat_obj,
                            features = gene,
                            reduction = "umap",
                            cols = rev(brewer.pal(n = 11, name = "RdBu")))
      print(plot_cl)
      
    }
    
    if (choice == "cond") {
      
       plot_cond = FeaturePlot(object = seurat_obj,
                            features = gene,
                            reduction = "umap",
                            split.by = "orig.ident",
                            keep.scale = "feature",
                            combine = FALSE,
                            cols = rev(brewer.pal(n = 11, name = "RdBu"))) 
    
      print(plot_grid(plotlist = plot_cond, ncol = 4))
 
    }
    
    if (choice != "clust" & choice != "cond") {
      
      print("Wrong parameter value.")
      
    }

  }
  
}

```

```{r, results="hold"}

# Global metabolic genes
plot_gene_fun(merged_prep, intersect(list_genes_metabo, rownames(merged_prep)), "clust") 
# Glutaminolysis genes
plot_gene_fun(merged_prep, intersect(list_genes_metabo_glut, rownames(merged_prep)), "clust") 
# Glycolysis genes
plot_gene_fun(merged_prep, intersect(list_genes_metabo_glycolysis, rownames(merged_prep)), "clust")
# TCA genes
plot_gene_fun(merged_prep, intersect(list_genes_metabo_tca, rownames(merged_prep)), "clust") 
# Hematopoietic Transcription Factors
plot_gene_fun(merged_prep, intersect(list_genes_TH_hemato, rownames(merged_prep)), "clust") 

```

```{r, results="hold", fig.width=14, fig.height=10}

# Global metabolic genes
plot_gene_fun(merged_prep, intersect(list_genes_metabo, rownames(merged_prep)), "cond") 
# Glutaminolysis genes
plot_gene_fun(merged_prep, intersect(list_genes_metabo_glut, rownames(merged_prep)), "cond") 
# Glycolysis genes
plot_gene_fun(merged_prep, intersect(list_genes_metabo_glycolysis, rownames(merged_prep)), "cond")
# TCA genes
plot_gene_fun(merged_prep, intersect(list_genes_metabo_tca, rownames(merged_prep)), "cond") 
# Hematopoietic Transcription Factors
plot_gene_fun(merged_prep, intersect(list_genes_TH_hemato, rownames(merged_prep)), "cond") 

```

```{r, fig.width = 14, fig.height = 10}

# Cluster only 
Idents(merged_prep) = "clusters"

p1 = DotPlot(merged_prep,
        features = list_genes_metabo,
        cols = plasma(2),
        dot.scale = 8) +
  RotatedAxis()

p2 = DotPlot(merged_prep,
        features = list_genes_metabo_glut,
        cols = plasma(2),
        dot.scale = 8) +
  RotatedAxis()

p3 = DotPlot(merged_prep,
        features = list_genes_metabo_glycolysis,
        cols = plasma(2),
        dot.scale = 8) +
  RotatedAxis()

p4 = DotPlot(merged_prep,
        features = list_genes_metabo_tca,
        cols = plasma(2),
        dot.scale = 8) +
  RotatedAxis()

p5 = DotPlot(merged_prep,
        features = list_genes_TH_hemato,
        cols = plasma(2),
        dot.scale = 8) +
  RotatedAxis()

# Conditions only
Idents(merged_prep) = "orig.ident"

p6 = DotPlot(merged_prep,
        features = list_genes_metabo,
        cols = plasma(2),
        dot.scale = 8) +
  RotatedAxis()

p7 = DotPlot(merged_prep,
        features = list_genes_metabo_glut,
        cols = plasma(2),
        dot.scale = 8) +
  RotatedAxis()

p8 = DotPlot(merged_prep,
        features = list_genes_metabo_glycolysis,
        cols = plasma(2),
        dot.scale = 8) +
  RotatedAxis()

p9 = DotPlot(merged_prep,
        features = list_genes_metabo_tca,
        cols = plasma(2),
        dot.scale = 8) +
  RotatedAxis()

p10 = DotPlot(merged_prep,
        features = list_genes_TH_hemato,
        cols = plasma(2),
        dot.scale = 8) +
  RotatedAxis()

# Clusters and conditions
Idents(merged_prep) = "orig.ident"

p11 = DotPlot(merged_prep,
        features = list_genes_metabo,
        cols = plasma(18),
        dot.scale = 8,
        split.by = "clusters") +
  RotatedAxis()

p12 = DotPlot(merged_prep,
        features = list_genes_metabo_glut,
        cols = plasma(18),
        dot.scale = 8,
        split.by = "clusters") +
  RotatedAxis()

p13 = DotPlot(merged_prep,
        features = list_genes_metabo_glycolysis,
        cols = plasma(18),
        dot.scale = 8,
        split.by = "clusters") +
  RotatedAxis()

p14 = DotPlot(merged_prep,
        features = list_genes_metabo_tca,
        cols = plasma(18),
        dot.scale = 8,
        split.by = "clusters") +
  RotatedAxis()

p15 = DotPlot(merged_prep,
        features = list_genes_TH_hemato,
        cols = c("red"),
        dot.scale = 8,
        split.by = "clusters") +
  RotatedAxis()

(p1 | p2 | p3) / (p4 | p5)
(p6 | p7 | p8) / (p9 | p10)

```

```{r, fig.width = 14, fig.height = 34}

(p11 | p12 | p13) / (p14 | p15)

```

```{r}

ggsave(plot = (p1 | p2 | p3) / (p4 | p5), width = 20, height = 16,
       filename = paste0(directory_output, "SCPA_dotplot_list_genes_clust.svg"))
ggsave(plot = (p6 | p7 | p8) / (p9 | p10), width = 20, height = 16,
       filename = paste0(directory_output, "SCPA_dotplot_list_genes_cond.svg"))
ggsave(plot = p11 | p12, width = 20, height = 22,
       filename = paste0(directory_output, "SCPA_dotplot_list_genes_clust_cond_1.svg"))
ggsave(plot = p13 | p14, width = 20, height = 22,
       filename = paste0(directory_output, "SCPA_dotplot_list_genes_clust_cond_2.svg"))
ggsave(plot = p15, width = 10, height = 22,
       filename = paste0(directory_output, "SCPA_dotplot_list_genes_clust_cond_3.svg"))

```

```{r, fig.width = 18, fig.height = 6}

v1 = VlnPlot(merged_prep,
             features = list_genes_metabo,
             cols = plasma(18),
             group.by = "orig.ident",
             split.by = "clusters",
             ncol = 1, 
             combine = FALSE)

v2 = VlnPlot(merged_prep,
             features = list_genes_metabo_glut,
             cols = plasma(18),
             group.by = "orig.ident",
             split.by = "clusters",
             ncol = 1, 
             combine = FALSE)

v3 = VlnPlot(merged_prep,
             features = list_genes_metabo_glycolysis,
             cols = plasma(18),
             group.by = "orig.ident",
             split.by = "clusters",
             ncol = 1, 
             combine = FALSE)

v4 = VlnPlot(merged_prep,
             features = list_genes_metabo_tca,
             cols = plasma(18),
             group.by = "orig.ident",
             split.by = "clusters",
             ncol = 1, 
             combine = FALSE)

v5 = VlnPlot(merged_prep,
             features = list_genes_TH_hemato,
             cols = plasma(18),
             group.by = "orig.ident",
             split.by = "clusters",
             ncol = 1, 
             combine = FALSE)

lapply(v1, function(x) { print(x) })
lapply(v2, function(x) { print(x) })
lapply(v3, function(x) { print(x) })
lapply(v4, function(x) { print(x) })
lapply(v5, function(x) { print(x) })

```


```{r, Rsession}

end_time = Sys.time()
cat("Total execution time : ", as.numeric (end_time - start_time, units = "mins"), "minutes")

# Show package version
sessionInfo()

```

```{r, results='hide'}

# Clean working space and memory 
rm(list = ls())
gc()

```

